package Bio::KBase::GenomeAnnotation::GenomeAnnotationImpl;
use strict;
use Bio::KBase::Exceptions;
# Use Semantic Versioning (2.0.0-rc.1)
# http://semver.org 
our $VERSION = "0.1.0";

=head1 NAME

GenomeAnnotation

=head1 DESCRIPTION

API Access to the Genome Annotation Service.

Provides support for gene calling, functional annotation, re-annotation. Use to extract annotation in
formation about an existing genome, or to create new annotations.

=cut

#BEGIN_HEADER

use Safe;
use File::Temp;
use File::Slurp;
use Data::Dumper;
use Digest::MD5 'md5_hex';
use Time::HiRes 'gettimeofday';
use POSIX;
use SOAP::Lite;

use Bio::KBase::IDServer::Client;
use Bio::KBase::KmerAnnotationByFigfam::Client;
use KmerClassifier;
#use Bio::KBase::KIDL::Helpers qw(json_to_tempfile tempfile_to_json);
use IPC::Run qw(run);
use JSON::XS;
use File::Slurp;
use Bio::KBase::GenomeAnnotation::Awe;
use Bio::KBase::GenomeAnnotation::Shock;

use Bio::KBase::GenomeAnnotation::Glimmer;
use GenomeTypeObject;
use gjogenbank;
use GenBankToGTO;
use IDclient;
use ANNOserver;
use SeedUtils;
use SeedURLs;
use gjoseqlib;
use StrepRepeats;
use overlap_resolution;
use PropagateGBMetadata;
use Capture::Tiny 'capture_stderr';
use AdaboostClassify;
use MongoDB;

use Bio::KBase::DeploymentConfig;

sub _get_coder
{
    return JSON::XS->new->ascii->pretty->allow_nonref;
}

sub _call_using_strep_repeats
{
    my($self, $genome_in, $tool) = @_;
    
    $genome_in = GenomeTypeObject->initialize($genome_in);
    
    my $parsed = StrepRepeats::get_raw_repeats($genome_in, $tool);

    my $event = {
	tool_name => $tool,
	execution_time => scalar gettimeofday,
	parameters => [],
	hostname => $self->{hostname},
    };

    my $idc = IDclient->new($genome_in);
    my $event_id = $genome_in->add_analysis_event($event);

    my $type = 'repeat_unit';

    for my $r (@$parsed)
    {
	my $loc = $r->{location};
	$genome_in->add_feature({
	    -id_client 	     => $idc,
	    -id_prefix 	     => $genome_in->{id},
	    -type 	     => $type,
	    -location 	     => $loc,
	    -function 	     => $r->{function},
	    -annotation      => $r->{annotation},
	    -analysis_event_id 	     => $event_id,
	});
    }

    $genome_in = $genome_in->prepare_for_return();

    return $genome_in;
}

sub _allocate_local_genome_id
{
    my($self, $taxon_id, $mongo_host, $mongo_db) = @_;

    my $query_timeout = -1;
    my $connection_timeout = 120;

    $mongo_db //= "seed-genome-allocation";
 
    my $client = MongoDB->connect($mongo_host);
    #my $client = MongoDB::MongoClient->new(host => $mongo_host, query_timeout => $query_timeout,
#					   timeout => $connection_timeout,
#					   auto_reconnect => 1,
#					   auto_connect => 1);
#
    my $db = $client->get_database($mongo_db);
    my $coll_next = $db->get_collection('next');

    $coll_next->ensure_index({prefix => 1});

    my $res = $coll_next->find_one_and_update({ taxon => $taxon_id }, { '$inc' => { next_val => 1 } }, { upsert => 1, new => 1 });
    #print Dumper($res);

    if (!ref($res))
    {
	die "MongoDB error: $res";
    }

    my $val = $res->{next_val};

    return "$taxon_id.0$val";
}


sub _allocate_seed_genome_id
{
    my($self, $taxon_id, $url) = @_;

    my $proxy = SOAP::Lite->uri('http://www.soaplite.com/Scripts')-> proxy($url);

    #
    # If we have heavy registration traffic we may fail with messages like
    #
    #   Error registering genome via SEED clearinghouse: soap:Server error executing queries: DBD::mysql::db do failed: Deadlock found when trying to get lock; try restarting transaction at /vol/core-seed/Clearinghouse/FIG/CGI/clearinghouse_services.cgi line 162.
    #
    # So we will do a backoff and retry here.
    #

    my $max_retries = 10;
    my $try = 0;
    while ($try++ < $max_retries)
    {
	my $r = $proxy->register_genome($taxon_id);
	if ($r->fault) {
	    warn "Error on try $try registering genome via SEED clearinghouse: " . $r->faultcode . " " . $r->faultstring;
	    sleep(2 + rand(10));
	}
	else
	{
	    my $id = $r->result;
	    return "$taxon_id.$id";
	}
    }
    #
    # Retries failed
    #
    die "Unable to register genome via SEED clearinghouse after $max_retries tries\n";
}

sub _allocate_kb_genome_id
{
    my($self, $taxon_id, $url) = @_;

    my $id_prefix = "kb";
	
    my $idc = Bio::KBase::IDServer::Client->new($url);
    my $id = "$id_prefix|g." . $idc->allocate_id_range("$id_prefix|g", 1);
    return $id;
}

sub _allocate_genome_id_fallback
{
    my($self, $taxon_id) = @_;

    my $id = "random|$taxon_id." . int(rand(10000));
    return $id;
}

#END_HEADER

sub new
{
    my($class, @args) = @_;
    my $self = {
    };
    bless $self, $class;
    #BEGIN_CONSTRUCTOR

    my $cfg = Bio::KBase::DeploymentConfig->new($ENV{KB_SERVICE_NAME} || "GenomeAnnotation");

    my $dir = $cfg->setting("kmer_v2_data_directory");
    $dir or die "Configuration parameter for kmer_v2_data_directory not set";
    -d $dir or die "Directory $dir for kmer_v2_data_directory does not exist";
	
    $self->{kmer_v2_data_directory} = $dir;

    if (my $temp = $cfg->setting("tempdir"))
    {
	$ENV{TEMPDIR} = $ENV{TMPDIR} = $temp;
    }

    my $dir = $cfg->setting("kmer_classifier_data_directory");
    #
    # Make these soft errors for now.
    #
    if (!$dir)
    {
	warn "Configuration parameter for kmer_classifier_data_directory not set";
    }
    elsif (! -d $dir)
    {
	warn "Directory $dir for kmer_classifier_data_directory does not exist";
    }
	
    $self->{kmer_classifier_data_directory} = $dir;

    $dir = $cfg->setting("nr-annotation-directory");
    if (!$dir)
    {
	warn "Configuration parameter for nr-annotation-directory not set";
    }
    elsif (! -d $dir)
    {
	warn "Directory $dir for nr-annotation_directory does not exist";
    }
	
    $self->{nr_annotation_directory} = $dir;

    my $phage_files = $cfg->setting("phage-annotation-files");
    if (!$phage_files)
    {
	$phage_files = [];
    }
    elsif (!ref($phage_files))
    {
	$phage_files = [$phage_files];
    }
    $self->{phage_annotation_files} = $phage_files;

    $self->{special_protein_threads} = $cfg->setting("special_protein_threads") // 1;
    $self->{special_protein_dbdir} = $cfg->setting("special_protein_dbdir");
    $self->{special_protein_lookup_db} = $cfg->setting("special_protein_lookup_db");
    $self->{special_protein_cache_db} = $cfg->setting("special_protein_cache_db");
    $self->{special_protein_cache_dbhost} = $cfg->setting("special_protein_cache_dbhost");
    $self->{special_protein_cache_dbuser} = $cfg->setting("special_protein_cache_dbuser");
    $self->{special_protein_cache_dbpass} = $cfg->setting("special_protein_cache_dbpass");

    $self->{patric_call_proteins_remote_host} = $cfg->setting("patric_call_proteins_remote_host");
    $self->{patric_call_proteins_remote_user} = $cfg->setting("patric_call_proteins_remote_user");
    $self->{patric_call_proteins_remote_key} = $cfg->setting("patric_call_proteins_remote_key");
    $self->{patric_call_proteins_path} = $cfg->setting("patric_call_proteins_path");
    $self->{patric_call_proteins_ff_path} = $cfg->setting("patric_call_proteins_ff_path");
    $self->{patric_call_proteins_md5_to_fam_path} = $cfg->setting("patric_call_proteins_md5_to_fam_path");

    $self->{patric_annotate_families_url} = $cfg->setting("patric_annotate_families_url");
    $self->{patric_annotate_families_kmers} = $cfg->setting("patric_annotate_families_kmers");

    $self->{mongo_allocate_id_host} = $cfg->setting("mongo-allocate-id-host");
    $self->{mongo_allocate_id_db} = $cfg->setting("mongo-allocate-id-db");

    $self->{patric_mlst_dbdir} = $cfg->setting("patric_mlst_dbdir");
    $self->{workflow_dir} = $cfg->setting("workflow-dir");

    #
    # String to use in CDS identifiers.
    #
    $self->{cds_id_type} = $cfg->setting("cds_id_type") || "CDS";

    #
    # Determine if we are using KB or SEED genome ID generation.
    #

    my $surl = $cfg->setting("seed-id-clearinghouse");
    my $iurl = $cfg->setting("idserver_url");

    #
    # For backward compatibility set this
    #
    $iurl = 'https://kbase.us/services/idserver' if !defined($iurl);

    if ($self->{mongo_allocate_id_host})
    {
	print STDERR "Using local mongo for id allocation\n";
	$self->{allocate_genome_id} = sub {
	    my($taxon_id) = @_;
	    $self->_allocate_local_genome_id($taxon_id, $self->{mongo_allocate_id_host}, $self->{mongo_allocate_id_db});
	};
    }
    elsif ($surl)
    {
	print STDERR "Using SEED id allocation from $surl\n";
	$self->{allocate_genome_id} = sub { my($taxon_id) = @_;
					    $self->_allocate_seed_genome_id($taxon_id, $surl);
					};
    }
    elsif ($iurl)
    {
	print STDERR "Using KBase id allocation from $iurl\n";
	$self->{allocate_genome_id} = sub { my($taxon_id) = @_;
					    $self->_allocate_kb_genome_id($taxon_id, $iurl);
					};
    }
    else
    {
	print STDERR "No id allocation mechanism defined\n";
	$self->{allocate_genome_id} = sub { my($taxon_id) = @_;
					    $self->_allocate_genome_id_fallback($taxon_id);
					};
    }
    
    $self->{kmer_service_url} = $cfg->setting("kmer_service_url");
    $self->{awe_server} = $cfg->setting("awe-server");
    $self->{shock_server} = $cfg->setting("shock-server");
    $self->{genemark_home} = $cfg->setting("genemark-home");

    print STDERR "kmer_v2_data_directory = $self->{kmer_v2_data_directory}\n";

    my $h = `hostname`;
    chomp $h;
    $self->{hostname} = $h;
    
    #END_CONSTRUCTOR

    if ($self->can('_init_instance'))
    {
	$self->_init_instance();
    }
    return $self;
}

=head1 METHODS



=head2 genome_ids_to_genomes

  $genomes = $obj->genome_ids_to_genomes($ids)

=over 4

=item Parameter and return types

=begin html

<pre>
$ids is a reference to a list where each element is a genome_id
$genomes is a reference to a list where each element is a genomeTO
genome_id is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$ids is a reference to a list where each element is a genome_id
$genomes is a reference to a list where each element is a genomeTO
genome_id is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given one or more Central Store genome IDs, convert them into genome objects.

=back

=cut

sub genome_ids_to_genomes
{
    my $self = shift;
    my($ids) = @_;

    my @_bad_arguments;
    (ref($ids) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"ids\" (value was \"$ids\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to genome_ids_to_genomes:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genome_ids_to_genomes');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genomes);
    #BEGIN genome_ids_to_genomes

    

    #END genome_ids_to_genomes
    my @_bad_returns;
    (ref($genomes) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"genomes\" (value was \"$genomes\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to genome_ids_to_genomes:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genome_ids_to_genomes');
    }
    return($genomes);
}




=head2 create_genome

  $genome = $obj->create_genome($metadata)

=over 4

=item Parameter and return types

=begin html

<pre>
$metadata is a genome_metadata
$genome is a genomeTO
genome_metadata is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	taxonomy has a value which is a string
	owner has a value which is a string
genome_id is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$metadata is a genome_metadata
$genome is a genomeTO
genome_metadata is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	taxonomy has a value which is a string
	owner has a value which is a string
genome_id is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Create a new genome object and assign metadata.

=back

=cut

sub create_genome
{
    my $self = shift;
    my($metadata) = @_;

    my @_bad_arguments;
    (ref($metadata) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"metadata\" (value was \"$metadata\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to create_genome:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome);
    #BEGIN create_genome

    $genome = GenomeTypeObject->new();
    $genome->set_metadata($metadata);

    #
    # If no ID was created, allocate a genome ID from our ID server.
    #
    if (!$genome->{id})
    {
	my $tax = $genome->{ncbi_taxonomy_id};
	
	if ($tax !~ /^\d+$/)
	{
	    $tax = "6666666";
	}

	$genome->{id} = $self->{allocate_genome_id}->($tax);
    }
    $genome = $genome->prepare_for_return();

    #END create_genome
    my @_bad_returns;
    (ref($genome) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome\" (value was \"$genome\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to create_genome:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome');
    }
    return($genome);
}




=head2 create_genome_from_genbank

  $genome = $obj->create_genome_from_genbank($gb_data)

=over 4

=item Parameter and return types

=begin html

<pre>
$gb_data is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$gb_data is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Create a new genome object from one or more genbank files.

=back

=cut

sub create_genome_from_genbank
{
    my $self = shift;
    my($gb_data) = @_;

    my @_bad_arguments;
    (!ref($gb_data)) or push(@_bad_arguments, "Invalid type for argument \"gb_data\" (value was \"$gb_data\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to create_genome_from_genbank:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_genbank');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome);
    #BEGIN create_genome_from_genbank

    #
    # We need to parse the file into a list of entries first so that
    # we may extract the taxon ID and allocate a genome ID. This way
    # our initial feature set can be created with the appropriate identifiers.
    #

    my @entries = parse_genbank(\$gb_data);
    
    my $tax_id;
    for my $entry (@entries)
    {
	my @sources = gjogenbank::features_of_type( $entry, 'source' );
	my ( $src_taxid ) = map { /^taxon:(\S+)$/ ? $1 : () }
		map { $_->[1]->{db_xref} ? @{$_->[1]->{db_xref}} : () }
		@sources;
	if ($src_taxid)
	{
	    $tax_id = $src_taxid;
	    last;
	}
    }

    if ($tax_id !~ /^\d+$/)
    {
	$tax_id = "6666666";
    }
    
    my $genome_id = $self->{allocate_genome_id}->($tax_id);

    $genome = GenBankToGTO::new({ entry => \@entries, id => $genome_id });

    #
    # The current genbank code is creating features with type peg; remap them to CDS.
    #

    for my $feature ($genome->features)
    {
	if ($feature->{type} eq 'peg')
	{
	    $feature->{type} = 'CDS';
	}
    }

    $genome->prepare_for_return();

    #END create_genome_from_genbank
    my @_bad_returns;
    (ref($genome) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome\" (value was \"$genome\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to create_genome_from_genbank:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_genbank');
    }
    return($genome);
}




=head2 create_genome_from_SEED

  $genome = $obj->create_genome_from_SEED($genome_id)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_id is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_id is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Create a new genome object based on data from the SEED project.

=back

=cut

sub create_genome_from_SEED
{
    my $self = shift;
    my($genome_id) = @_;

    my @_bad_arguments;
    (!ref($genome_id)) or push(@_bad_arguments, "Invalid type for argument \"genome_id\" (value was \"$genome_id\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to create_genome_from_SEED:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_SEED');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome);
    #BEGIN create_genome_from_SEED
    #END create_genome_from_SEED
    my @_bad_returns;
    (ref($genome) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome\" (value was \"$genome\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to create_genome_from_SEED:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_SEED');
    }
    return($genome);
}




=head2 create_genome_from_RAST

  $genome = $obj->create_genome_from_RAST($genome_or_job_id)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_or_job_id is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_or_job_id is a string
$genome is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Create a new genome object based on a RAST genome.

=back

=cut

sub create_genome_from_RAST
{
    my $self = shift;
    my($genome_or_job_id) = @_;

    my @_bad_arguments;
    (!ref($genome_or_job_id)) or push(@_bad_arguments, "Invalid type for argument \"genome_or_job_id\" (value was \"$genome_or_job_id\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to create_genome_from_RAST:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_RAST');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome);
    #BEGIN create_genome_from_RAST

    print STDERR "get RAST : ctx=" . Dumper($ctx);
    $genome = {};
    #END create_genome_from_RAST
    my @_bad_returns;
    (ref($genome) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome\" (value was \"$genome\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to create_genome_from_RAST:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'create_genome_from_RAST');
    }
    return($genome);
}




=head2 set_metadata

  $genome_out = $obj->set_metadata($genome_in, $metadata)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$metadata is a genome_metadata
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

genome_metadata is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	taxonomy has a value which is a string
	owner has a value which is a string

</pre>

=end html

=begin text

$genome_in is a genomeTO
$metadata is a genome_metadata
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

genome_metadata is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	taxonomy has a value which is a string
	owner has a value which is a string


=end text



=item Description

Modify genome metadata.

=back

=cut

sub set_metadata
{
    my $self = shift;
    my($genome_in, $metadata) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($metadata) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"metadata\" (value was \"$metadata\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to set_metadata:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'set_metadata');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN set_metadata

    $genome_out = GenomeTypeObject->initialize_without_indexes($genome_in);
    $genome_out->set_metadata($metadata);
    $genome_out = $genome_out->prepare_for_return();
    
    #end set_metadata
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to set_metadata:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'set_metadata');
    }
    return($genome_out);
}




=head2 add_contigs

  $genome_out = $obj->add_contigs($genome_in, $contigs)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
contig_id is a string
bool is an int
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) an int
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	closeness_measure has a value which is a float
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string

</pre>

=end html

=begin text

$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
contig_id is a string
bool is an int
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) an int
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	closeness_measure has a value which is a float
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string


=end text



=item Description

Add a set of contigs to the genome object.

=back

=cut

sub add_contigs
{
    my $self = shift;
    my($genome_in, $contigs) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($contigs) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"contigs\" (value was \"$contigs\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to add_contigs:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_contigs');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #END set_metadata
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to set_metadata:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'set_metadata');
    }
    return($genome_out);
}




=head2 add_contigs

  $genome_out = $obj->add_contigs($genome_in, $contigs)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Add a set of contigs to the genome object.

=back

=cut

sub add_contigs
{
    my $self = shift;
    my($genome_in, $contigs) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($contigs) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"contigs\" (value was \"$contigs\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to add_contigs:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_contigs');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN add_contigs

    $genome_out = GenomeTypeObject->initialize_without_indexes($genome_in);
    $genome_out->add_contigs($contigs);
    $genome_out = $genome_out->prepare_for_return();

    #END add_contigs
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to add_contigs:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_contigs');
    }
    return($genome_out);
}




=head2 add_contigs_from_handle

  $genome_out = $obj->add_contigs_from_handle($genome_in, $contigs)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$contigs is a reference to a list where each element is a contig
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Add a set of contigs to the genome object, loading the contigs
from the given handle service handle.

=back

=cut

sub add_contigs_from_handle
{
    my $self = shift;
    my($genome_in, $contigs) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($contigs) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"contigs\" (value was \"$contigs\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to add_contigs_from_handle:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_contigs_from_handle');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN add_contigs_from_handle
    #END add_contigs_from_handle
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to add_contigs_from_handle:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_contigs_from_handle');
    }
    return($genome_out);
}




=head2 add_features

  $genome_out = $obj->add_features($genome_in, $features)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$features is a reference to a list where each element is a compact_feature
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

compact_feature is a reference to a list containing 5 items:
	0: (id) a string
	1: (location) a string
	2: (feature_type) a string
	3: (function) a string
	4: (aliases) a string

</pre>

=end html

=begin text

$genome_in is a genomeTO
$features is a reference to a list where each element is a compact_feature
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

compact_feature is a reference to a list containing 5 items:
	0: (id) a string
	1: (location) a string
	2: (feature_type) a string
	3: (function) a string
	4: (aliases) a string


=end text



=item Description

Add a set of features in tabular form.

=back

=cut

sub add_features
{
    my $self = shift;
    my($genome_in, $features) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($features) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"features\" (value was \"$features\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to add_features:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_features');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN add_features

    $genome_out = GenomeTypeObject->initialize($genome_in);
    $genome_out->add_features_from_list($features);
    $genome_out = $genome_out->prepare_for_return();
    
    #END add_features
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to add_features:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'add_features');
    }
    return($genome_out);
}




=head2 genomeTO_to_reconstructionTO

  $return = $obj->genomeTO_to_reconstructionTO($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a reconstructionTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a reconstructionTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string


=end text



=item Description



=back

=cut

sub genomeTO_to_reconstructionTO
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to genomeTO_to_reconstructionTO:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genomeTO_to_reconstructionTO');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN genomeTO_to_reconstructionTO

    use ANNOserver;
    my $annoO = ANNOserver->new;
    my %in_models = map { chop; ($_ => 1) } `all_roles_used_in_models`;
    my %bindings_to_roles;

    my $features = $genomeTO->{features};
    my @role_fid_tuples;
    my $assignments = [];
    foreach my $fidH (@$features)
    {
	my $fid = $fidH->{id};
	my $f = $fidH->{function};
	if ($f)
	{
	    
	    push(@$assignments,[$fid,$f]);
	    foreach my $role (&SeedUtils::roles_of_function($f))
	    {
		push(@role_fid_tuples,[$role,$fid]);
		if ($in_models{$role}) { $bindings_to_roles{$role}->{$fid} = 1 }
	    }
	}
    }
    my $mr = $annoO->metabolic_reconstruction({-roles => \@role_fid_tuples});
    my $sub_vars = [];
    my $bindings = [];
    my %subsys;
    foreach my $tuple (@$mr)
    {
	my($sub_var,$role,$fid) = @$tuple;
	my($sub,$var) = split(/:/,$sub_var);
	if ($var !~ /\*?(0|-1)\b/)
	{
	    $subsys{$sub} = $var;
	    $bindings_to_roles{$role}->{$fid} = 1;
	}
    }
    foreach my $role (keys(%bindings_to_roles))
    {
	my $roles = $bindings_to_roles{$role};
	my @fids = keys(%$roles);
	foreach my $fid (@fids)
	{
	    push(@$bindings,[$fid,$role]);
	}
    }
    my @sv = map { [$_,$subsys{$_}] } keys(%subsys);
    $return = {
	subsystems => \@sv,
	assignments => $assignments,
	bindings   => $bindings,
    };
    
    #END genomeTO_to_reconstructionTO
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to genomeTO_to_reconstructionTO:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genomeTO_to_reconstructionTO');
    }
    return($return);
}




=head2 genomeTO_to_feature_data

  $return = $obj->genomeTO_to_feature_data($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a fid_data_tuples
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

fid_data_tuples is a reference to a list where each element is a fid_data_tuple
fid_data_tuple is a reference to a list containing 4 items:
	0: a fid
	1: a md5
	2: a location
	3: a function
fid is a string
md5 is a string
function is a string

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a fid_data_tuples
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

fid_data_tuples is a reference to a list where each element is a fid_data_tuple
fid_data_tuple is a reference to a list containing 4 items:
	0: a fid
	1: a md5
	2: a location
	3: a function
fid is a string
md5 is a string
function is a string


=end text



=item Description



=back

=cut

sub genomeTO_to_feature_data
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to genomeTO_to_feature_data:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genomeTO_to_feature_data');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN genomeTO_to_feature_data

    my $feature_data = [];
    my $features = $genomeTO->{features};
    foreach my $feature (@$features)
    {
	my $fid = $feature->{id};
	my $loc = join(",",map { my($contig,$beg,$strand,$len) = @$_; 
				 "$contig\_$beg$strand$len" 
			       } @{$feature->{location}});
	my $type = $feature->{type};
	my $func = $feature->{function};
	my $md5 = "";
	$md5 = md5_hex(uc($feature->{protein_translation})) if $feature->{protein_translation};
	my $aliases = join(",",@{$feature->{aliases}});
	push(@$feature_data,[$fid,$loc,$type,$func,$aliases,$md5]);
    }
    $return = $feature_data;
    #END genomeTO_to_feature_data
    my @_bad_returns;
    (ref($return) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to genomeTO_to_feature_data:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'genomeTO_to_feature_data');
    }
    return($return);
}




=head2 reconstructionTO_to_roles

  $return = $obj->reconstructionTO_to_roles($reconstructionTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$reconstructionTO is a reconstructionTO
$return is a reference to a list where each element is a role
reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string

</pre>

=end html

=begin text

$reconstructionTO is a reconstructionTO
$return is a reference to a list where each element is a role
reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string


=end text



=item Description



=back

=cut

sub reconstructionTO_to_roles
{
    my $self = shift;
    my($reconstructionTO) = @_;

    my @_bad_arguments;
    (ref($reconstructionTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"reconstructionTO\" (value was \"$reconstructionTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to reconstructionTO_to_roles:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'reconstructionTO_to_roles');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN reconstructionTO_to_roles

    my $bindings = $reconstructionTO->{bindings};
    my %roles = map { ($_->[1] => 1) } @$bindings;
    $return = [sort keys(%roles)];

    #END reconstructionTO_to_roles
    my @_bad_returns;
    (ref($return) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to reconstructionTO_to_roles:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'reconstructionTO_to_roles');
    }
    return($return);
}




=head2 reconstructionTO_to_subsystems

  $return = $obj->reconstructionTO_to_subsystems($reconstructionTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$reconstructionTO is a reconstructionTO
$return is a variant_subsystem_pairs
reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string

</pre>

=end html

=begin text

$reconstructionTO is a reconstructionTO
$return is a variant_subsystem_pairs
reconstructionTO is a reference to a hash where the following keys are defined:
	subsystems has a value which is a variant_subsystem_pairs
	bindings has a value which is a fid_role_pairs
	assignments has a value which is a fid_function_pairs
variant_subsystem_pairs is a reference to a list where each element is a variant_of_subsystem
variant_of_subsystem is a reference to a list containing 2 items:
	0: a subsystem
	1: a variant
subsystem is a string
variant is a string
fid_role_pairs is a reference to a list where each element is a fid_role_pair
fid_role_pair is a reference to a list containing 2 items:
	0: a fid
	1: a role
fid is a string
role is a string
fid_function_pairs is a reference to a list where each element is a fid_function_pair
fid_function_pair is a reference to a list containing 2 items:
	0: a fid
	1: a function
function is a string


=end text



=item Description



=back

=cut

sub reconstructionTO_to_subsystems
{
    my $self = shift;
    my($reconstructionTO) = @_;

    my @_bad_arguments;
    (ref($reconstructionTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"reconstructionTO\" (value was \"$reconstructionTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to reconstructionTO_to_subsystems:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'reconstructionTO_to_subsystems');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN reconstructionTO_to_subsystems

    my $subsys_pairs = $reconstructionTO->{subsystems};
    $return = $subsys_pairs;
    
    #END reconstructionTO_to_subsystems
    my @_bad_returns;
    (ref($return) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to reconstructionTO_to_subsystems:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'reconstructionTO_to_subsystems');
    }
    return($return);
}




=head2 assign_functions_to_CDSs

  $return = $obj->assign_functions_to_CDSs($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome object populated with contig data, perform gene calling
and functional annotation and return the annotated genome.

=back

=cut

sub assign_functions_to_CDSs
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to assign_functions_to_CDSs:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'assign_functions_to_CDSs');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN assign_functions_to_CDSs
    my $features = $genomeTO->{features};
    my %to;
    my $i;
    my @prots;
    for ($i=0; ($i < @$features); $i++)
    {
	$to{$features->[$i]->{id}} = $i;
	my $fid = $features->[$i];
	my $translation;
	if (defined($translation = $fid->{protein_translation}))
	{
	    my $id = $fid->{id};
	    push(@prots,[$id,'',$translation]);
	}
    }
    my $anno = ANNOserver->new();
    my $handle = $anno->assign_function_to_prot(-input => \@prots,
						-kmer => 8,
						-scoreThreshold => 3,
						-seqHitThreshold => 3);
    while (my $res = $handle->get_next())
    {
	my($id, $function, $otu, $score, $nonoverlap_hits, $overlap_hits, $details, $fam) = @$res;
	$features->[$to{$id}]->{function} = $function;
	push(@{$features->[$to{$id}]->{annotations}},
	     ["Set function to\n$function\nby assign_function_to_CDSs with otu=$otu score=$score nonoverlap=$nonoverlap_hits hits=$overlap_hits figfam=$fam",
	      'genome annotation service',
	      time
	     ]);
    }
    $return = $genomeTO;
    #END assign_functions_to_CDSs
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to assign_functions_to_CDSs:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'assign_functions_to_CDSs');
    }
    return($return);
}




=head2 annotate_genome

  $return = $obj->annotate_genome($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_genome
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_genome:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_genome');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN annotate_genome
    my $genome = $genomeTO;
    my $anno = ANNOserver->new();

    if ($genome->{genetic_code} !~ /^\d+$/)
    {
	die "Genome has invalid genetic code $genome->{genetic_code}";
    }
    if ($genome->{scientific_name} eq '')
    {
	die "Genome does not have a scientific name defined";
    }
    if ($genome->{domain} eq '')
    {
	die "Genome does not have a domain defined";
    }

    #
    # Reformat the contigs for use with the ANNOserver.
    #
    my @contigs;
    foreach my $gctg (@{$genome->{contigs}})
    {
	push(@contigs, [$gctg->{id}, undef, $gctg->{dna}]);
    }

    #
    # Call genes.
    #
    print STDERR "Call genes...\n";
    my $peg_calls = $anno->call_genes(-input => \@contigs, -geneticCode => $genome->{genetic_code});
    print STDERR "Call genes...done\n";


    #
    # Call RNAs
    #
    my($genus, $species, $strain) = split(/\s+/, $genome->{scientific_name}, 3);
    $species = "sp" if $species eq '';
    print STDERR "Call rnas '$genus' '$species' '$strain' '$genome->{domain}'...\n";
    my $rna_calls = $anno->find_rnas(-input => \@contigs, -genus => $genus, -species => $species,
				     -domain => $genome->{domain});
    print STDERR "Call rnas...done\n";

    my($fasta_rna, $rna_locations) = @$rna_calls;

    my %feature_loc;
    my %feature_func;
    my %feature_anno;
    
    for my $ent (@$rna_locations)
    {
	my($loc_id, $contig, $start, $stop, $func) = @$ent;
	my $len = abs($stop - $start) + 1;
	my $strand = ($stop > $start) ? '+' : '-';
	$feature_loc{$loc_id} = [$contig, $start, $strand, $len];
	$feature_func{$loc_id} = $func if $func;
    }

    my($fasta_proteins, $protein_locations) = @$peg_calls;

    my $features = $genome->{features};
    if (!$features)
    {
	$features = [];
	$genome->{features} = $features;
    }

    #
    # Assign functions for proteins.
    #

    my $prot_fh;
    open($prot_fh, "<", \$fasta_proteins) or die "Cannot open the fasta string as a filehandle: $!";
    my $handle = $anno->assign_function_to_prot(-input => $prot_fh,
						-kmer => 8,
						-scoreThreshold => 3,
						-seqHitThreshold => 3);
    while (my $res = $handle->get_next())
    {
	my($id, $function, $otu, $score, $nonoverlap_hits, $overlap_hits, $details, $fam) = @$res;
	$feature_func{$id} = $function;
	$feature_anno{$id} = "Set function to\n$function\nby assign_function_to_prot with otu=$otu score=$score nonoverlap=$nonoverlap_hits hits=$overlap_hits figfam=$fam";
    }
    close($prot_fh);
    
    for my $ent (@$protein_locations)
    {
	my($loc_id, $contig, $start, $stop) = @$ent;
	my $len = abs($stop - $start) + 1;
	my $strand = ($stop > $start) ? '+' : '-';
	$feature_loc{$loc_id} = [$contig, $start, $strand, $len];
    }

    my $id_server = IDclient->new($genome);

    #
    # Create features for PEGs
    #
    my $n_pegs = @$protein_locations;
    my $protein_prefix = "$genome->{id}.CDS";
    my $peg_id_start = $id_server->allocate_id_range($protein_prefix, $n_pegs) + 0;
    print STDERR "allocated CDS id start $peg_id_start for $n_pegs CDSs\n";

    open($prot_fh, "<", \$fasta_proteins) or die "Cannot open the fasta string as a filehandle: $!";
    my $next_id = $peg_id_start;
    while (my($id, $def, $seq) = read_next_fasta_seq($prot_fh))
    {
	my $loc = $feature_loc{$id};
	my $kb_id = "$protein_prefix.$next_id";
	$next_id++;
	my $annos = [];
	push(@$annos, ['Initial gene call performed by call_genes',
		       'genome annotation service',
		       time
		       ]);
	if ($feature_anno{$id})
	{
	    push(@$annos, [$feature_anno{$id}, 'genome annotation service', time]);
	}
	my $feature = {
	    id => $kb_id,
	    location => [$loc],
	    type => 'CDS',
	    protein_translation => $seq,
	    aliases => [],
	    $feature_func{$id} ? (function => $feature_func{$id}) : (),
	    annotations => $annos,
	};
	push(@$features, $feature);
    }
    close($prot_fh);

    #
    # Create features for RNAs
    #
    my $n_rnas = @$rna_locations;
    my $rna_prefix = "$genome->{id}.rna";
    my $rna_id_start = $id_server->allocate_id_range($rna_prefix, $n_rnas) + 0;
    print STDERR "allocated id start $rna_id_start for $n_rnas nras\n";

    my $rna_fh;
    open($rna_fh, "<", \$fasta_rna) or die "Cannot open the fasta string as a filehandle: $!";
    $next_id = $rna_id_start;
    while (my($id, $def, $seq) = read_next_fasta_seq($rna_fh))
    {
	my $loc = $feature_loc{$id};
	my $kb_id = "$rna_prefix.$next_id";
	$next_id++;
	my $feature = {
	    id => $kb_id,
	    location => [$loc],
	    type => 'rna',
	    $feature_func{$id} ? (function => $feature_func{$id}) : (),
	    aliases => [],
	    annotations => [ ['Initial RNA call performed by find_rnas', 'genome annotation service', time] ],
	};
	push(@$features, $feature);
    }

    $return = $genome;
    
    #END annotate_genome
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_genome:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_genome');
    }
    return($return);
}




=head2 call_selenoproteins

  $return = $obj->call_selenoproteins($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_selenoproteins
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_selenoproteins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_selenoproteins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_selenoproteins

    my $idc = IDclient->new($genomeTO);

    my $coder = _get_coder();
    
    my $genomeTO_json = $coder->encode($genomeTO);

    my $genomeOut_json;

    my $tmp = File::Temp->new();
    print $tmp $genomeTO_json;
    close($tmp);

    my @cmd = ('rast_call_special_proteins',
	       '--seleno',
	       '--input', $tmp);
    $ctx->stderr->log_cmd(@cmd);
    my $ok = run(\@cmd,
		 '>', \$genomeOut_json,
		 $ctx->stderr->redirect);

    undef $tmp;
    undef $genomeTO;

    if ($ok)
    {
	$return = $coder->decode($genomeOut_json);
    }
    else
    {
	die "rast_call_special_proteins failed: ", $ctx->stderr->text_value;
    }
    
    #END call_selenoproteins
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_selenoproteins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_selenoproteins');
    }
    return($return);
}




=head2 call_pyrrolysoproteins

  $return = $obj->call_pyrrolysoproteins($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_pyrrolysoproteins
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_pyrrolysoproteins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_pyrrolysoproteins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_pyrrolysoproteins
    
    my $coder = _get_coder();
    
    my $genomeTO_json = $coder->encode($genomeTO);

    my $genomeOut_json;

    my $tmp = File::Temp->new();
    print $tmp $genomeTO_json;
    close($tmp);

    my @cmd = ('rast_call_special_proteins',
	       '--pyrro',
	       '--input', $tmp);
    $ctx->stderr->log_cmd(@cmd);
			  
    my $ok = run(\@cmd,
		 '>', \$genomeOut_json,
		 $ctx->stderr->redirect);

    undef $tmp;
    undef $genomeTO;

    if ($ok)
    {
	$return = $coder->decode($genomeOut_json);
    }
    else
    {
	die "rast_call_special_proteins failed: " . $ctx->stderr->text_value;
    }

    #END call_pyrrolysoproteins
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_pyrrolysoproteins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_pyrrolysoproteins');
    }
    return($return);
}




=head2 call_features_selenoprotein

  $return = $obj->call_features_selenoprotein($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome typed object, call selenoprotein features.

=back

=cut

sub call_features_selenoprotein
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_selenoprotein:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_selenoprotein');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_selenoprotein

    $return = $self->call_selenoproteins($genomeTO);
    
    #END call_features_selenoprotein
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_selenoprotein:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_selenoprotein');
    }
    return($return);
}




=head2 call_features_pyrrolysoprotein

  $return = $obj->call_features_pyrrolysoprotein($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome typed object, call pyrrolysoprotein features.

=back

=cut

sub call_features_pyrrolysoprotein
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_pyrrolysoprotein:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_pyrrolysoprotein');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_pyrrolysoprotein

    $return = $self->call_pyrrolysoproteins($genomeTO);
    
    #END call_features_pyrrolysoprotein
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_pyrrolysoprotein:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_pyrrolysoprotein');
    }
    return($return);
}




=head2 call_features_insertion_sequences

  $return = $obj->call_features_insertion_sequences($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome typed object, call insertion sequences.

=back

=cut

sub call_features_insertion_sequences
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_insertion_sequences:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_insertion_sequences');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_insertion_sequences

    my $genome_in = GenomeTypeObject->initialize($genomeTO);
    
    my $sequences_file = $genome_in->extract_contig_sequences_to_temp_file();
    my $output_file = File::Temp->new();

    my @params = ();
    my @cmd = ("find_IS", @params);
    $ctx->stderr->log_cmd(@cmd);

    my $ok = run(\@cmd,
		 "<", $sequences_file,
		 ">", $output_file,
		 $ctx->stderr->redirect);

    unlink($sequences_file);

    if (!$ok)
    {
	die "Error running find_IS: @cmd\n";
    }

    close($output_file);
    my($res_fh);
    open($res_fh, "<", $output_file) or die "Cannot open find_IS output file $output_file: $!";

    my $event = {
	tool_name => $cmd[0],
	execution_time => scalar gettimeofday,
	parameters => \@params,
	hostname => $self->{hostname},
    };

    my $idc = IDclient->new($genome_in);
    my $event_id = $genome_in->add_analysis_event($event);

    #	fig|NC_002952.IS.1	NC_002952_760426_758423	IS1182	IS1182 family insertion sequence
    #	fig|NC_002952.IS.2	NC_002952_2545636_2543556	IS_Unknown	Putative insertion sequence

    my $id_prefix = $genome_in->{id};

    while(<$res_fh>)
    {
	chomp;
	my($id, $loc, $type, $function) = split(/\t/);

	my $typed_prefix = join(".", $id_prefix, $type);

	my $cur_id_suffix = $idc->allocate_id_range($typed_prefix, 1);

	my $id = join(".", $typed_prefix, $cur_id_suffix);

	my $locs = GenomeTypeObject::seed_location_to_location_list($loc);

	$genome_in->add_feature({
	    -id              => $id,
	    -type 	     => $type,
	    -location 	     => $locs,
	    -function 	     => $function,
	    -analysis_event_id 	     => $event_id,
	});
    }

    $return = $genome_in;
    $return = $return->prepare_for_return();

    #END call_features_insertion_sequences
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_insertion_sequences:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_insertion_sequences');
    }
    return($return);
}




=head2 call_features_rRNA_SEED

  $genome_out = $obj->call_features_rRNA_SEED($genome_in, $types)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$types is a reference to a list where each element is a rna_type
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

rna_type is a string

</pre>

=end html

=begin text

$genome_in is a genomeTO
$types is a reference to a list where each element is a rna_type
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

rna_type is a string


=end text



=item Description

Given a genome typed object, find instances of ribosomal RNAs in
the genome.
The types parameter is used to select the types of RNAs to
call. It is a list of strings where each value is one of
   "5S"
   "SSU"
   "LSU"
or "ALL" to choose all available rRNA types.

=back

=cut

sub call_features_rRNA_SEED
{
    my $self = shift;
    my($genome_in, $types) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($types) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"types\" (value was \"$types\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_rRNA_SEED:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_rRNA_SEED');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_rRNA_SEED
    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my %types = map { lc($_) => 1 } @$types;

    my @type_args;

    if ($types{all} || @$types == 0)
    {
	# Don't need to set type arg since tool defaults to calling all.
    }
    else
    {
	for my $type (qw(5S SSU LSU))
	{
	    if ($types{lc($type)})
	    {
		push(@type_args, "-$type");
	    }
	}
    }

    my @cmd = ("rast_call_rRNAs", "--input", $tmp_in, "--output", $tmp_out,
	       "--id-prefix", $genome_in->{id}, @type_args);
    $ctx->stderr->log_cmd(@cmd);
    my $ok = run(\@cmd, $ctx->stderr->redirect);
    if (!$ok)
    {
	die "error calling rRNAs: $?\non command @cmd\n" . $ctx->stderr->text_value;
    }

    $genome_out = $coder->decode(scalar read_file("" . $tmp_out));
    #END call_features_rRNA_SEED
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_rRNA_SEED:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_rRNA_SEED');
    }
    return($genome_out);
}




=head2 call_features_tRNA_trnascan

  $genome_out = $obj->call_features_tRNA_trnascan($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome typed object, find instances of tRNAs in
the genome.

=back

=cut

sub call_features_tRNA_trnascan
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_tRNA_trnascan:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_tRNA_trnascan');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_tRNA_trnascan
    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @cmd = ("rast_call_tRNAs", "--input", $tmp_in, "--output", $tmp_out,
	       "--id-prefix", $genome_in->{id});

    $ctx->stderr->log_cmd(@cmd);

    my $ok = run(\@cmd, $ctx->stderr->redirect);

    if (!$ok)
    {
	die "error calling tRNAs: $?\non command @cmd\n" . $ctx->stderr->text_value;
    }

    $genome_out = $coder->decode(scalar read_file("" . $tmp_out));
    #END call_features_tRNA_trnascan
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_tRNA_trnascan:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_tRNA_trnascan');
    }
    return($genome_out);
}




=head2 call_RNAs

  $genome_out = $obj->call_RNAs($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Given a genome typed object, find instances of all RNAs we currently
have support for detecting.

=back

=cut

sub call_RNAs
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_RNAs:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_RNAs');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_RNAs
    ##################################################
    use ANNOserver;
    my $anno = ANNOserver->new();
    #
    # Reformat the contigs for use with the ANNOserver.
    #
    my @contigs;
    foreach my $gctg (@{$genome_in->{contigs}})
    {
	push(@contigs, [$gctg->{id}, undef, $gctg->{dna}]);
    }

    if ($genome_in->{scientific_name} eq '')
    {
	die "Genome does not have a scientific name defined";
    }
    if ($genome_in->{domain} eq '')
    {
	die "Genome does not have a domain defined";
    }

    #
    # Call RNAs
    #
    my($genus, $species, $strain) = split(/\s+/, $genome_in->{scientific_name}, 3);
    $species = "sp" if $species eq '';
    print STDERR "Call rnas '$genus' '$species' '$strain' '$genome_in->{domain}'...\n";
    my $rna_calls = $anno->find_rnas(-input => \@contigs, -genus => $genus, -species => $species,
				     -domain => $genome_in->{domain});
    print STDERR "Call rnas...done\n";
    my($fasta_rna, $rna_locations) = @$rna_calls;

    my %feature_loc;
    my %feature_func;
    my %feature_anno;
    
    for my $ent (@$rna_locations)
    {
	my($loc_id, $contig, $start, $stop, $func) = @$ent;
	my $len = abs($stop - $start) + 1;
	my $strand = ($stop > $start) ? '+' : '-';
	$feature_loc{$loc_id} = [$contig, $start, $strand, $len];
	$feature_func{$loc_id} = $func if $func;
    }
    my $features = $genome_in->{features};
    if (!$features)
    {
	$features = [];
	$genome_in->{features} = $features;
    }

    my $id_server = IDclient->new($genome_in);
    
    #
    # Create features for RNAs
    #
    my $n_rnas = @$rna_locations;
    my $rna_prefix = "$genome_in->{id}.rna";
    my $rna_id_start = $id_server->allocate_id_range($rna_prefix, $n_rnas) + 0;
    print STDERR "allocated id start $rna_id_start for $n_rnas nras\n";

    my $rna_fh;
    open($rna_fh, "<", \$fasta_rna) or die "Cannot open the fasta string as a filehandle: $!";
    my $next_id = $rna_id_start;
    while (my($id, $def, $seq) = read_next_fasta_seq($rna_fh))
    {
	my $loc = $feature_loc{$id};
	my $kb_id = "$rna_prefix.$next_id";
	$next_id++;
	my $feature = {
	    id => $kb_id,
	    location => [$loc],
	    type => 'rna',
	    $feature_func{$id} ? (function => $feature_func{$id}) : (),
	    aliases => [],
	    annotations => [ ['Initial RNA call performed by find_rnas', 'genome annotation service', time] ],
	};
	push(@$features, $feature);
    }
    $genome_out = $genome_in;

    #END call_RNAs
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_RNAs:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_RNAs');
    }
    return($genome_out);
}




=head2 call_features_CDS_glimmer3

  $return = $obj->call_features_CDS_glimmer3($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a glimmer3_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a glimmer3_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_CDS_glimmer3
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_CDS_glimmer3:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_glimmer3');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_CDS_glimmer3

    my $genome_in = GenomeTypeObject->initialize($genomeTO);
    my $sequences_file = $genome_in->extract_contig_sequences_to_temp_file();

    my $gparams = { %$params };
    $gparams->{genetic_code} = $genome_in->{genetic_code};
    my $calls;

    my $stderr = capture_stderr {
	$calls = Bio::KBase::GenomeAnnotation::Glimmer::call_genes_with_glimmer($sequences_file, $gparams, $ctx);
    };
    if (!$ctx->stderr->log($stderr))
    {
	print STDERR $stderr;
    }

    unlink($sequences_file);

    my $trans_table = SeedUtils::genetic_code($genome_in->{genetic_code});
    
    my $event = {
	tool_name => "glimmer3",
	execution_time => scalar gettimeofday,
	parameters => [ map { join("=", $_, $gparams->{$_}) } sort keys %$gparams ],
	hostname => $self->{hostname},
    };

    my $idc = IDclient->new($genome_in);
    
    my $event_id = $genome_in->add_analysis_event($event);
    my $type = 'CDS';
    my $id_type = $self->{cds_id_type};

    my $id_prefix = $genome_in->{id};
    if ($id_prefix =~ /^\d+\.\d+$/)
    {
	$id_prefix = "fig|$id_prefix";
    }
    my $typed_prefix = join(".", $id_prefix, $id_type);

    my $count = @$calls;
    my $cur_id_suffix = $idc->allocate_id_range($typed_prefix, $count);

    for my $call (@$calls)
    {
	my($fid, $contig, $begin, $end, $dna) = @$call;

	my($strand, $len);
	my $fix_start = 1;
	if ($begin < $end)
	{
	    $fix_start = 0 if $begin <= 3;

	    $strand = '+';
	    $len = $end - $begin + 1;
	}
	else
	{
	    my $cobj = $genome_in->find_contig($contig);
	    if (ref $cobj)
	    {
		my $clen = length($cobj->{dna});
		$fix_start = 0 if $begin > ($clen - 3);
	    }

	    $strand = '-';
	    $len = $begin - $end + 1;
	}

	my $loc = [[$contig, $begin, $strand, $len]];

	my $trans = SeedUtils::translate($dna, $trans_table, $fix_start);
	$trans =~ s/\*$//;

	my $id = join(".", $typed_prefix, $cur_id_suffix);
	$cur_id_suffix++;
	
	$genome_in->add_feature({
	    -id		     => $id,
	    -type 	     => $type,
	    -location 	     => $loc,
	    -analysis_event_id 	     => $event_id,
	    -annotator => 'glimmer3',
	    -protein_translation => $trans,
	});
    }
				
    $return = $genome_in;
    $return = $return->prepare_for_return();
    
    #END call_features_CDS_glimmer3
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_CDS_glimmer3:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_glimmer3');
    }
    return($return);
}




=head2 call_features_CDS_prodigal

  $return = $obj->call_features_CDS_prodigal($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_CDS_prodigal
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_CDS_prodigal:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_prodigal');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_CDS_prodigal

    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genomeTO));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my $stderr = $ctx->stderr;

    my $id_prefix = $genomeTO->{id};
    print "1 '$id_prefix'\n";
    if ($id_prefix =~ /^\d+\.\d+$/)
    {
	$id_prefix = "fig|$id_prefix";
       print "2 '$id_prefix'\n";
    }
    print "3 '$id_prefix'\n";

    my @cmd = ("rast_call_CDSs_using_prodigal", "--input", $tmp_in, "--output", $tmp_out,
	       "--id-prefix", $id_prefix, "--id-type", $self->{cds_id_type});
    $stderr->log(join(" ", @cmd));

    my $ok = run(\@cmd, $stderr->redirect);
    
    if (!$ok)
    {
	die "error calling CDSs: $?\non command @cmd\n" . $stderr->text_value . "\n";
    }

    $return = $coder->decode(scalar read_file("" . $tmp_out));
    
    #END call_features_CDS_prodigal
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_CDS_prodigal:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_prodigal');
    }
    return($return);
}




=head2 call_features_CDS_genemark

  $return = $obj->call_features_CDS_genemark($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_CDS_genemark
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_CDS_genemark:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_genemark');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_CDS_genemark

    #
    # Genemark requires a license file installed in the "home directory".
    # We expect a parameter to be passed that defines this directory.
    # We also assume that this home directory is the genemark code release. This
    # simplifies things since we can't freely distribute the code in a KB runtime anyway.
    #
    my $gmark_home = $self->{genemark_home};

    #
    # Set up the environment for our subprocess.
    #
    my %env = (HOME => $gmark_home);

    my $genome_in = GenomeTypeObject->initialize($genomeTO);
    my($gc) = $genome_in->compute_contigs_gc();

    $gc = int($gc + 0.5);
    print STDERR "gc=$gc\n";

    #
    # Genemark 3.2.5 supports gc 30-70 inclusive.
    #
    $gc = 30 if $gc < 30;
    $gc = 70 if $gc > 70;
    
    my $code = $genome_in->{genetic_code} || 11;
    if ($code != 11 && $code != 4)
    {
	die "Genetic code must be either 4 or 11 for genemark";
    }

    my $h_file = "heu_${code}_${gc}.mod";
    my $h_path = "$gmark_home/heuristic_mod/$h_file";

    if (! -f $h_path)
    {
	die "Heuristic file $h_path not found";
    }

    my $tmp_in = $genome_in->extract_contig_sequences_to_temp_file();
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @cmd = ("$gmark_home/gmhmmp",
	       "-f" => "G",
	       "-r",
	       "-a",
	       "-g", $code,
	       "-m", $h_path,
	       "-o", "$tmp_out",
	       $tmp_in);

    $ctx->stderr->log(join(" ", @cmd)) if $ctx;

    my @cmds = (\@cmd,
		init => sub {
		    $ENV{$_} = $env{$_} foreach keys %env;
		});

    my $event = {
	tool_name => "genemark",
	execution_time => scalar gettimeofday,
	parameters => \@cmd,
	hostname => $self->{hostname},
    };
    my $event_id = $genome_in->add_analysis_event($event);
    my $type = 'CDS';

    my $ok = run(@cmds, ($ctx ? $ctx->stderr->redirect : ()));

    if (!$ok)
    {
	die "Error running pipeline: @cmd\n" . $ctx->stderr->text_value . "\n" if $ctx;
    }

    my $fh;
    open($fh, "<", $tmp_out) or die "Cannot open $tmp_out: $!";

    #
    # Parse generated GFF.
    #
    my $l;
    while (defined($l = <$fh>))
    {
	last unless $l =~ /^#/;
    }

    my %by_gene;
    while (defined($l))
    {
	next if $l =~ /^\s*$/;
	next if $l =~ /^\#/;

	chomp $l;

	my(@fields) = split(/\t/, $l);

	@fields == 9 or die "Invalid GFF at line $.";
	
	my($ctg, $who, $type, $start, $end, $score, $strand, $frame, $attr) = @fields;

	my @attrs = split(/,\s*/, $attr);
	my %attrs = map { my @a = split(/=/, $_, 2); $a[0] => $a[1] } @attrs;

	my $loc;
	my $len = $end - $start + 1;
	if ($strand eq '+') {
	    $loc = [[ $ctg, $start, '+', $len]];
	} else {
	    $loc = [[ $ctg, $end, '-', $len]];
	}

	my $info = [-type => 'CDS',
		    -id_type => $self->{cds_id_type},
		    -location => $loc,
		    -annotator => $who,
		    -annotation => "Add feature called by $who using $h_file",
		    -analysis_event_id => $event_id,
		    -quality_measure => { genemark_score => $score },
		    ];
	$by_gene{$attrs{gene_id}} = $info;
    } continue {
	$l = <$fh>;
    }

    # print STDERR Dumper(\%by_gene);
   

    close($fh);
    open($fh, "<", $tmp_out) or die "Cannot open $tmp_out: $!";
    my $l;
    while (defined($l = <$fh>))
    {
	last unless $l =~ /^#/;
    }

    #
    # We've read features. Set up for creating the IDs etc, then read the proteins and
    # add features to the GTO.
    #
    
    my $id_prefix = $genome_in->{id};
    if ($id_prefix =~ /^\d+\.\d+$/)
    {
	$id_prefix = "fig|$id_prefix";
    }
    my $typed_prefix = join(".", $id_prefix, $self->{cds_id_type});

    my $count = int(%by_gene);
    my $idc = IDclient->new($genome_in);
    my $cur_id_suffix = $idc->allocate_id_range($typed_prefix, $count);

    my $cur;
    my $cur_type;
    my $cur_prot;
    while (defined($l)) {
	next if $l !~ /^\#\#/;
	    
	if ($l =~ /^\#\#(\S+)\s+(\S+)/)
	{
	    $cur_type = $1;
	    $cur = $2;
	    $cur_prot = '';
	    # print STDERR "Set cur_type=$cur_type  cur=$cur\n";
	}
	elsif ($l =~ /^\#\#end-(\S+)/)
	{
	    # print "End: 1=$1 cur=$cur cur_type=$cur_type\n";
	    if ($1 eq 'Protein' && $cur_type eq 'Protein')
	    {
		my $info = $by_gene{$cur};
		if (!$info)
		{
		    warn "No info for gene $info";
		    next;
		}

		my $id = join(".", $typed_prefix, $cur_id_suffix);
		$cur_id_suffix++;

		$genome_in->add_feature({
		    -id => $id,
		    -protein_translation => $cur_prot,
		    @$info,
		});
	    }
	    undef $cur_type;
	    undef $cur;
	    undef $cur_prot;
	}
	elsif ($l =~ /^\#\#(\S+)/)
	{
	    if ($cur_type eq 'Protein')
	    {
		$cur_prot .= $1;
	    }
	    else
	    {
		print STDERR "Skip '$cur_type' $l\n";
	    }
	}
    } continue {
	$l = <$fh>;
    }
    
    $return = $genome_in;
    $return->prepare_for_return();

    #END call_features_CDS_genemark
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_CDS_genemark:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_genemark');
    }
    return($return);
}




=head2 call_features_CDS_SEED_projection

  $return = $obj->call_features_CDS_SEED_projection($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a SEED_projection_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

SEED_projection_parameters is a reference to a hash where the following keys are defined:
	reference_database has a value which is a string
	reference_id has a value which is a string
	kmer_size has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a SEED_projection_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

SEED_projection_parameters is a reference to a hash where the following keys are defined:
	reference_database has a value which is a string
	reference_id has a value which is a string
	kmer_size has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_CDS_SEED_projection
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_CDS_SEED_projection:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_SEED_projection');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_CDS_SEED_projection

    #
    # Do some validity checking.
    #

    #
    # Default to coreseed.
    #
    $params->{reference_database} //= 'core';
    my $ref_db = $params->{reference_database};

    if ($ref_db eq 'PATRIC')
    {
	die "PATRIC reference database not yet supported\n";
    }

    if ($params->{reference_id} eq '')
    {
	die "No reference ID provided\n";
    }

    if ($ref_db !~ /^http/)
    {
	my $where = SeedURLs::url($ref_db);
	if (!$where)
	{
	    die "Could not find reference database '$ref_db'\n";
	}
    }

    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genomeTO));

    my @cmd = ("fast_project",
	       "--input", $tmp_in,
	       "-s", $ref_db,
	       "-r", $params->{reference_id},
	       (defined($params->{kmer_size}) ? ("--kmersize", $params->{kmer_size}) : ()),
	       "--format", "gto");

    $ctx->stderr->log_cmd(@cmd);
    my $genomeOut_json;
    my $ok = run(\@cmd,
		 '>', \$genomeOut_json,
		 $ctx->stderr->redirect);

    if (!$ok)
    {
	die "error calling fast_project: $?\non command @cmd";
    }

    $return = $coder->decode($genomeOut_json);


    #END call_features_CDS_SEED_projection
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_CDS_SEED_projection:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_SEED_projection');
    }
    return($return);
}




=head2 call_features_CDS_FragGeneScan

  $return = $obj->call_features_CDS_FragGeneScan($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_CDS_FragGeneScan
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_CDS_FragGeneScan:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_FragGeneScan');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_CDS_FragGeneScan
    die "Not implemented";
    #END call_features_CDS_FragGeneScan
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_CDS_FragGeneScan:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_CDS_FragGeneScan');
    }
    return($return);
}




=head2 call_features_repeat_region_SEED

  $genome_out = $obj->call_features_repeat_region_SEED($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is a repeat_region_SEED_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is a repeat_region_SEED_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_repeat_region_SEED
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_repeat_region_SEED:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_repeat_region_SEED');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_repeat_region_SEED

    $genome_in = GenomeTypeObject->initialize($genome_in);
    
    my $sequences_file = $genome_in->extract_contig_sequences_to_temp_file();
    my $output_file = File::Temp->new();

    my @opts;
    push(@opts, "-i", $params->{min_identity}) if defined($params->{min_identity});
    push(@opts, "-l", $params->{min_length}) if defined($params->{min_length});

    my $stderr = $ctx->stderr;

    my @cmd = ("svr_big_repeats", @opts);

    $stderr->log_cmd(@cmd);

    my $tmpdir = File::Temp->newdir(undef, CLEANUP => 1);

    print STDERR "seq file $sequences_file\n";
    my $ok = run(\@cmd,
		 init => sub { print STDERR "init chdir to $tmpdir\n"; chdir($tmpdir) or die $!; },
		 "<", $sequences_file,
		 "|",
		 ["svr_condense_repeats"],
		 ">", $output_file,
		 $stderr->redirect,
		);

    unlink($sequences_file);

    if (!$ok)
    {
	die "Error running svr_big_repeats: @cmd\n" . $stderr->text_value . "\n";
    }

    close($output_file);
    my($res_fh);
    open($res_fh, "<", $output_file) or die "Cannot open svr_big_repeats file $output_file: $!";

    my $event = {
	tool_name => "svr_big_repeats | svr_condense_repeats",
	execution_time => scalar gettimeofday,
	parameters => \@opts,
	hostname => $self->{hostname},
    };

    my $event_id = $genome_in->add_analysis_event($event);

    # olson@bio-data-1:~/FIGdisk/dist/releases/dev2$ svr_condense_repeats < r
    #    NC_000913       15377   16741   99.71
    #    NC_000913       19796   20564   98.83

    my $type = 'repeat';
    my $function = 'repeat region';

    while(<$res_fh>)
    {
	chomp;
	my($contig, $left, $right, $iden) = split(/\t/);

	next unless $left =~ /^\d+$/ && $right =~ /^\d+$/;

	my $confidence = $iden / 100.0;

	my $quality = {
	    existence_confidence => $confidence,
	};

	my $len = 1 + $right - $left;
	my $loc = [[$contig, $left, '+', $len]];
	$genome_in->add_feature({
	    -id_prefix 	     => $genome_in->{id},
	    -type 	     => $type,
	    -location 	     => $loc,
	    -function 	     => $function,
	    -analysis_event_id 	     => $event_id,
	    -quality_measure => $quality,
	});
    }

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END call_features_repeat_region_SEED
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_repeat_region_SEED:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_repeat_region_SEED');
    }
    return($genome_out);
}




=head2 call_features_prophage_phispy

  $genome_out = $obj->call_features_prophage_phispy($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_prophage_phispy
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_prophage_phispy:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_prophage_phispy');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_prophage_phispy
    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @cmd = ("rast_call_prophage_using_phispy", "--input", $tmp_in, "--output", $tmp_out,
	       "--id-prefix", $genome_in->{id});

    my $stderr = $ctx->stderr;
    $stderr->log_cmd(@cmd);

    my $ok = run(\@cmd, $stderr->redirect);

    if (!$ok)
    {
	die "error calling prophages: $?\non command @cmd\n" . $stderr->text_value . "\n";
    }

    $genome_out = $coder->decode(scalar read_file("" . $tmp_out));
    #END call_features_prophage_phispy
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_prophage_phispy:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_prophage_phispy');
    }
    return($genome_out);
}




=head2 call_features_scan_for_matches

  $genome_out = $obj->call_features_scan_for_matches($genome_in, $pattern, $feature_type)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$pattern is a string
$feature_type is a string
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$pattern is a string
$feature_type is a string
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_scan_for_matches
{
    my $self = shift;
    my($genome_in, $pattern, $feature_type) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (!ref($pattern)) or push(@_bad_arguments, "Invalid type for argument \"pattern\" (value was \"$pattern\")");
    (!ref($feature_type)) or push(@_bad_arguments, "Invalid type for argument \"feature_type\" (value was \"$feature_type\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_scan_for_matches:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_scan_for_matches');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_scan_for_matches
    die "Not implemented";
    #END call_features_scan_for_matches
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_scan_for_matches:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_scan_for_matches');
    }
    return($genome_out);
}




=head2 call_features_assembly_gap

  $genome_out = $obj->call_features_assembly_gap($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is an assembly_gap_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

assembly_gap_parameters is a reference to a hash where the following keys are defined:
	min_gap_length has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is an assembly_gap_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

assembly_gap_parameters is a reference to a hash where the following keys are defined:
	min_gap_length has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_assembly_gap
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_assembly_gap:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_assembly_gap');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_assembly_gap
    #END call_features_assembly_gap
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_assembly_gap:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_assembly_gap');
    }
    return($genome_out);
}




=head2 annotate_proteins_similarity

  $return = $obj->annotate_proteins_similarity($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a similarity_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a similarity_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description

Annotate based on similarity to annotation databases.

=back

=cut

sub annotate_proteins_similarity
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_proteins_similarity:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_similarity');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN annotate_proteins_similarity

    my $coder = _get_coder();
    my $stderr = $ctx->stderr;
    
    my $dir = $self->{nr_annotation_directory};
    if (! -d $dir)
    {
	warn "NR annotation directory '$dir' empty or missing\n";
	$return = $genomeTO;
    }
    else
    {
	my $genomeTO_json = $coder->encode($genomeTO);

	my $genomeOut_json;

	my $tmp = File::Temp->new();
	print $tmp $genomeTO_json;
	close($tmp);

	print STDERR "Starting $dir $tmp\n";
	my @cmd = ('rast_annotate_proteins_similarity',
		   "--nr-dir", $dir,
		   ($params->{annotate_hypothetical_only} ? ("-H") : ()),
		   '--input', $tmp);
	my $ok = run(\@cmd,
		     '>', \$genomeOut_json,
		     $stderr->redirect);

	undef $tmp;
	undef $genomeTO;

	if ($ok) {
	    $return = $coder->decode($genomeOut_json);
	} else {
	    die "rast_annotate_proteins_similarity failed ($?):\n" . $stderr->text_value;
	}
    }

    #END annotate_proteins_similarity
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_proteins_similarity:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_similarity');
    }
    return($return);
}




=head2 annotate_proteins_phage

  $return = $obj->annotate_proteins_phage($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a phage_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

phage_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a phage_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

phage_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description

Annotate based on similarity to the phage annotation daatabase.

=back

=cut

sub annotate_proteins_phage
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_proteins_phage:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_phage');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN annotate_proteins_phage

    my $coder = _get_coder();
    my $stderr = $ctx->stderr;
    
    my $files = $self->{phage_annotation_files};
    if (!ref($files) || @$files == 0)
    {
	warn "No phage annotation files found\n";
	$return = $genomeTO;
    }
    else
    {
	my $tmpcur = File::Temp->new();
	print $tmpcur $coder->encode($genomeTO);
	close($tmpcur);

	undef $genomeTO;

	for my $pfile (@$files)
	{
	    my $tmpout = File::Temp->new();
	    close($tmpout);
		
	    print STDERR "Starting phage annotation $pfile < $tmpcur > $tmpout\n";
	    my @cmd = ('rast_annotate_proteins_similarity',
		       "--nr-file", $pfile,
		       ($params->{annotate_hypothetical_only} ? ("-H") : ()),
		       '--input', $tmpcur,
		       '--output', $tmpout);
	    
	    my $ok = run(\@cmd,
			 $stderr->redirect);

	    if (!$ok)
	    {
		die "rast_annotate_proteins_similarity failed during phage annotation ($?) cmd=@cmd:\n" . $stderr->text_value;
	    }

	    undef $tmpcur;
	    $tmpcur = $tmpout;
	}

	$return = $coder->decode(scalar read_file("" . $tmpcur));
    }

    #END annotate_proteins_phage
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_proteins_phage:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_phage');
    }
    return($return);
}




=head2 annotate_proteins_kmer_v1

  $return = $obj->annotate_proteins_kmer_v1($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a kmer_v1_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a kmer_v1_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub annotate_proteins_kmer_v1
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_proteins_kmer_v1:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_kmer_v1');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN annotate_proteins_kmer_v1

    my $n_proteins_per_call = 100;

    my $genome_in = GenomeTypeObject->initialize($genomeTO);

    my $kmer_service = Bio::KBase::KmerAnnotationByFigfam::Client->new($self->{kmer_service_url});


    if (!defined($params->{dataset_name}))
    {
	$params->{dataset_name} = $kmer_service->get_default_dataset_name();
    }

    if (!defined($params->{kmer_size}))
    {
	$params->{kmer_size} = 8;
    }

    warn "Annotating using dataset $params->{dataset_name} kmer_size $params->{kmer_size}\n";
    my $event = {
	tool_name => "KmerAnnotationByFigfam",
	execution_time => scalar gettimeofday,
	parameters => [ map { join("=", $_, $params->{$_}) } sort keys %$params ],
	hostname => $self->{hostname},
    };

    my $event_id = $genome_in->add_analysis_event($event);

    my @proteins;

    my $do_anno = sub {
	my($proteins) = @_;
	my $res = $kmer_service->annotate_proteins($proteins, $params);
	for my $hit (@$res)
	{
	    my($id, $func, $otu, $score, $nonover, $over, $details) = @$hit;
	    if ($func)
	    {
		$genome_in->update_function("annotate_proteins_kmer_v1", $id, $func, $event_id);
		my $feature = $genome_in->find_feature($id);
		if (ref($feature) eq 'HASH')
		{
		    $feature->{ quality }->{ hit_count } = $score;
		}
	    }
	}
    };

    for my $feature ($genome_in->features)
    {
	my $trans = $feature->{protein_translation};
	next unless $trans;

	if ($params->{annotate_hypothetical_only})
	{
	    my $f = $feature->{function};
	    if ($f ne '' && $f !~ /^\s*hypothetical\s+protein\s*$/i)
	    {
		next;
	    }
	}

	push(@proteins, [$feature->{id}, $trans]);
	if (@proteins >= $n_proteins_per_call)
	{
	    $do_anno->(\@proteins);
	    @proteins = ();
	}
    }
    $do_anno->(\@proteins) if @proteins;
    
    $return = $genome_in->prepare_for_return();

    #END annotate_proteins_kmer_v1
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_proteins_kmer_v1:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_kmer_v1');
    }
    return($return);
}




=head2 annotate_proteins_kmer_v2

  $genome_out = $obj->annotate_proteins_kmer_v2($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is a kmer_v2_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is a kmer_v2_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub annotate_proteins_kmer_v2
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_proteins_kmer_v2:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_kmer_v2');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_proteins_kmer_v2

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $filter;
    if ($params->{annotate_hypothetical_only})
    {
	$filter = sub {
	    my($feature) = @_;
	    my $f = $feature->{function};
	    if ($f eq '' || $f =~ /^\s*hypothetical\s+protein\s*$/i)
	    {
		return 1;
	    }
	    else
	    {
		return 0;
	    }
	};
    }
    my $sequences_file = $genome_in->extract_protein_sequences_to_temp_file($filter);
    my $output_file = File::Temp->new();

    my $min_hits = 5;
    my $max_gap = 200;

    if (defined($params->{min_hits}))
    {
	$min_hits = $params->{min_hits};
    }
	
    if (defined($params->{max_gap}))
    {
	$max_gap = $params->{max_gap};
    }

    my @params = ("-a", "-g", $max_gap, "-m", $min_hits, "-d", $self->{kmer_v2_data_directory});

    #
    # If we're configured with a families server and it is configured
    # with our v2 kmers, use it for function assignment.
    #
    if (($self->{patric_annotate_families_kmers} eq $self->{kmer_v2_data_directory}) &&
	$self->{patric_annotate_families_url})
    {
	push(@params, "-u", $self->{patric_annotate_families_url} . "/query");
    }

    my @cmd = ("kmer_search", @params);
    $ctx->stderr->log_cmd(@cmd);
    my $ok = run(\@cmd,
		 "<", $sequences_file,
		 ">", $output_file,
		 $ctx->stderr->redirect);
    
    unlink($sequences_file);

    if (!$ok)
    {
	die "Error running kmer_search ($?): @cmd\n" . $ctx->stderr->text_value;
    }

    close($output_file);
    my($res_fh);
    open($res_fh, "<", $output_file) or die "Cannot open kmer_search output file $output_file: $!";

    my $event = {
	tool_name => "kmer_search",
	execution_time => scalar gettimeofday,
	parameters => \@params,
	hostname => $self->{hostname},
    };

    my $event_id = $genome_in->add_analysis_event($event);
    
    while(<$res_fh>)
    {
	chomp;
	#  my($fid, $function) = split(/\t/);
	my($fid, $function, $hits, $hitsW) = split(/\t/);
	$genome_in->update_function("annotate_proteins_kmer_v2", $fid, $function, $event_id);

	my $feature = $genome_in->find_feature($fid);
	if (ref($feature) eq 'HASH')
	{
	    $feature->{ quality }->{ hit_count } = $hits;
	    $feature->{ quality }->{ weighted_hit_count } = $hitsW;
	}
    }

    $genome_out = $genome_in->prepare_for_return();

    #END annotate_proteins_kmer_v2
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_proteins_kmer_v2:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins_kmer_v2');
    }
    return($genome_out);
}




=head2 resolve_overlapping_features

  $genome_out = $obj->resolve_overlapping_features($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is a resolve_overlapping_features_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

resolve_overlapping_features_parameters is a reference to a hash where the following keys are defined:
	placeholder has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is a resolve_overlapping_features_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

resolve_overlapping_features_parameters is a reference to a hash where the following keys are defined:
	placeholder has a value which is an int


=end text



=item Description



=back

=cut

sub resolve_overlapping_features
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to resolve_overlapping_features:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'resolve_overlapping_features');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN resolve_overlapping_features

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $stderr;

    $ctx->stderr->log_cmd("resolve_overlapping_features", Dumper($params));
    $stderr = capture_stderr {
	$genome_out = overlap_resolution::resolve_overlapping_features($genome_in, $params);
    };
    if (!$ctx->stderr->log($stderr))
    {
	print STDERR $stderr;
    }

    $genome_out = $genome_out->prepare_for_return();

    #END resolve_overlapping_features
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to resolve_overlapping_features:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'resolve_overlapping_features');
    }
    return($genome_out);
}




=head2 propagate_genbank_feature_metadata

  $genome_out = $obj->propagate_genbank_feature_metadata($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is a propagate_genbank_feature_metadata_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

propagate_genbank_feature_metadata_parameters is a reference to a hash where the following keys are defined:
	min_rna_pct_coverage has a value which is a float

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is a propagate_genbank_feature_metadata_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

propagate_genbank_feature_metadata_parameters is a reference to a hash where the following keys are defined:
	min_rna_pct_coverage has a value which is a float


=end text



=item Description



=back

=cut

sub propagate_genbank_feature_metadata
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to propagate_genbank_feature_metadata:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'propagate_genbank_feature_metadata');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN propagate_genbank_feature_metadata

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $stderr;

    $ctx->stderr->log_cmd("propagate_genbank_feature_metadata", Dumper($params));
    $stderr = capture_stderr {
	$genome_out = PropagateGBMetadata::propagate_gb_metadata($genome_in, $params, $ctx->user_id);

    };
    if (!$ctx->stderr->log($stderr))
    {
	print STDERR $stderr;
    }

    $genome_out = $genome_out->prepare_for_return();

    #END propagate_genbank_feature_metadata
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to propagate_genbank_feature_metadata:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'propagate_genbank_feature_metadata');
    }
    return($genome_out);
}




=head2 call_features_ProtoCDS_kmer_v1

  $return = $obj->call_features_ProtoCDS_kmer_v1($genomeTO, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$params is a kmer_v1_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genomeTO is a genomeTO
$params is a kmer_v1_parameters
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_ProtoCDS_kmer_v1
{
    my $self = shift;
    my($genomeTO, $params) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_ProtoCDS_kmer_v1:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_ProtoCDS_kmer_v1');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_ProtoCDS_kmer_v1

    my $genome_in = GenomeTypeObject->initialize($genomeTO);
    
    my $min_hits = 5;
    my $max_gap = 200;

    if (defined($params->{min_hits}))
    {
	$min_hits = $params->{min_hits};
    }
	
    if (defined($params->{max_gap}))
    {
	$max_gap = $params->{max_gap};
    }

    my $event = {
	tool_name => "KmerAnnotationByFigfam",
	execution_time => scalar gettimeofday,
	parameters => [ map { join("=", $_, $params->{$_}) } sort keys %$params ],
	hostname => $self->{hostname},
    };

    my $idc = IDclient->new($genome_in);
    
    my $event_id = $genome_in->add_analysis_event($event);

    my $type = 'protoCDS';
    my $id_prefix = $genome_in->{id};
    if ($id_prefix =~ /^\d+\.\d+$/)
    {
	$id_prefix = "fig|$id_prefix";
    }
    my $typed_prefix = join(".", $id_prefix, $type);

    my $kmer_service = Bio::KBase::KmerAnnotationByFigfam::Client->new($self->{kmer_service_url});
    if (!defined($params->{dataset_name}))
    {
	$params->{dataset_name} = $kmer_service->get_default_dataset_name();
    }

    if (!defined($params->{kmer_size}))
    {
	$params->{kmer_size} = 8;
    }
    
    for my $ctg ($genome_in->contigs)
    {
	my $hits = $kmer_service->call_genes_in_dna([[$ctg->{id}, $ctg->{dna}]], $params);
	# print STDERR Dumper($hits);

	my $count = @$hits;
	my $cur_id_suffix = $idc->allocate_id_range($typed_prefix, $count);

	for my $hit (@$hits)
	{
	    my($nhits, $id, $begin, $end, $function, $otu) = @$hit;
	    my $quality = {
		existence_confidence => 0.5,
		hit_count => 0 + $nhits,
	    };
	    my($strand, $len);
	    if ($begin < $end)
	    {
		$strand = '+';
		$len = $end - $begin + 1;
	    }
	    else
	    {
		$strand = '-';
		$len = $begin - $end + 1;
	    }
	    
	    my $fid = join(".", $typed_prefix, $cur_id_suffix);
	    $cur_id_suffix++;

	    my $loc = [[$id, $begin, $strand, $len]];
	    $genome_in->add_feature({
		-id		     => $fid,
		-type 	     => $type,
		-location 	     => $loc,
		-function 	     => $function,
		-analysis_event_id 	     => $event_id,
		-quality_measure => $quality,
	    });
	}
    }

    $return = $genome_in;
    $return = $return->prepare_for_return();

    #END call_features_ProtoCDS_kmer_v1
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_ProtoCDS_kmer_v1:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_ProtoCDS_kmer_v1');
    }
    return($return);
}




=head2 call_features_ProtoCDS_kmer_v2

  $genome_out = $obj->call_features_ProtoCDS_kmer_v2($genome_in, $params)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$params is a kmer_v2_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$params is a kmer_v2_parameters
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub call_features_ProtoCDS_kmer_v2
{
    my $self = shift;
    my($genome_in, $params) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($params) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"params\" (value was \"$params\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_ProtoCDS_kmer_v2:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_ProtoCDS_kmer_v2');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_ProtoCDS_kmer_v2

    $genome_in = GenomeTypeObject->initialize($genome_in);
    
    my $sequences_file = $genome_in->extract_contig_sequences_to_temp_file();
    my $output_file = File::Temp->new();

    my $min_hits = 5;
    my $max_gap = 200;

    if (defined($params->{min_hits}))
    {
	$min_hits = $params->{min_hits};
    }
	
    if (defined($params->{max_gap}))
    {
	$max_gap = $params->{max_gap};
    }

    my @params = ("-g", $max_gap, "-m", $min_hits, "-d", $self->{kmer_v2_data_directory});

    my @cmd = ("kmer_search", @params);
    my $ok = run(\@cmd,
		 "<", $sequences_file,
		 ">", $output_file);

    unlink($sequences_file);

    if (!$ok)
    {
	die "Error running kmer_search: @cmd\n";
    }

    close($output_file);
    my($res_fh);
    open($res_fh, "<", $output_file) or die "Cannot open kmer_search output file $output_file: $!";

    my $event = {
	tool_name => "kmer_search",
	execution_time => scalar gettimeofday,
	parameters => \@params,
	hostname => $self->{hostname},
    };

    my $idc = IDclient->new($genome_in);
    my $event_id = $genome_in->add_analysis_event($event);

#	NC_002952       2902278 2902407 -       1       37      LSU ribosomal protein L34p      116.145798
#	Contig
#	Left
#	Right
#	Strand
#	Frame
#	Score-1
#	Function
#	Score-2
#    

    my $type = 'protoCDS';
    my $id_prefix = $genome_in->{id};
    if ($id_prefix =~ /^\d+\.\d+$/)
    {
	$id_prefix = "fig|$id_prefix";
    }
    my $typed_prefix = join(".", $id_prefix, $type);

    my @hits;

    while(<$res_fh>)
    {
	chomp;
	my($contig, $left, $right, $strand, $frame, $hit_count, $function, $weighted_hit_count) = split(/\t/, $_);

	next unless $left =~ /^\d+$/ && $right =~ /^\d+$/;

	push(@hits, $_);
    }
    close($res_fh);

    my $count = @hits;
    my $cur_id_suffix = $idc->allocate_id_range($typed_prefix, $count);

    for my $hit (@hits)
    {
	my($contig, $left, $right, $strand, $frame, $hit_count, $function, $weighted_hit_count) = split(/\t/, $hit);

	my $confidence = 1 - 0.5 ** ($weighted_hit_count / 3);

	my $quality = {
	    existence_confidence => $confidence,
	    hit_count => 0 + $hit_count,
	    weighted_hit_count => 0 + $weighted_hit_count,
	};

	my $id = join(".", $typed_prefix, $cur_id_suffix);
	$cur_id_suffix++;

	my $begin = 0 + (($strand eq '+') ? $left : $right);
	my $len = 1 + $right - $left;
	my $loc = [[$contig, $begin, $strand, $len]];
	$genome_in->add_feature({
	    -id              => $id,
	    -type 	     => $type,
	    -location 	     => $loc,
	    -function 	     => $function,
	    -analysis_event_id 	     => $event_id,
	    -quality_measure => $quality,
	});
    }

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END call_features_ProtoCDS_kmer_v2
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_ProtoCDS_kmer_v2:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_ProtoCDS_kmer_v2');
    }
    return($genome_out);
}




=head2 enumerate_special_protein_databases

  $database_names = $obj->enumerate_special_protein_databases()

=over 4

=item Parameter and return types

=begin html

<pre>
$database_names is a reference to a list where each element is a string

</pre>

=end html

=begin text

$database_names is a reference to a list where each element is a string


=end text



=item Description



=back

=cut

sub enumerate_special_protein_databases
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($database_names);
    #BEGIN enumerate_special_protein_databases

    $database_names = [];

    my $dir = $self->{special_protein_dbdir};
    if ($dir && opendir(my $dh, $dir))
    {
	my @list = map { s/\.faa$//; $_ } grep { -f "$dir/$_" && /\.faa$/ } readdir($dh);
	push(@$database_names, @list);
	closedir($dh);
    }

    #END enumerate_special_protein_databases
    my @_bad_returns;
    (ref($database_names) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"database_names\" (value was \"$database_names\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to enumerate_special_protein_databases:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'enumerate_special_protein_databases');
    }
    return($database_names);
}




=head2 compute_special_proteins

  $results = $obj->compute_special_proteins($genome_in, $database_names)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$database_names is a reference to a list where each element is a string
$results is a reference to a list where each element is a special_protein_hit
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

special_protein_hit is a reference to a list containing 7 items:
	0: (protein_id) a string
	1: (database_name) a string
	2: (database_id) a string
	3: (protein_coverage) a string
	4: (database_coverage) a string
	5: (identity) a float
	6: (p_value) a float

</pre>

=end html

=begin text

$genome_in is a genomeTO
$database_names is a reference to a list where each element is a string
$results is a reference to a list where each element is a special_protein_hit
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

special_protein_hit is a reference to a list containing 7 items:
	0: (protein_id) a string
	1: (database_name) a string
	2: (database_id) a string
	3: (protein_coverage) a string
	4: (database_coverage) a string
	5: (identity) a float
	6: (p_value) a float


=end text



=item Description



=back

=cut

sub compute_special_proteins
{
    my $self = shift;
    my($genome_in, $database_names) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($database_names) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"database_names\" (value was \"$database_names\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to compute_special_proteins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'compute_special_proteins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($results);
    #BEGIN compute_special_proteins

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $dir = $self->{special_protein_dbdir};
    
    my $prots = File::Temp->new();
    close($prots);
    $genome_in->write_protein_translations_to_file($prots);

    my $out = File::Temp->new();

    my @dbargs = map { ("--db", $_) } @$database_names;

    if (!$self->{special_protein_cache_db})
    {
	push(@dbargs, "--no-cache");
    }

    my @cmd = ('rast_compute_specialty_genes',
	       "--in", $prots,
	       "--db-dir", $dir,
	       @dbargs,
	       $self->{special_protein_cache_db} ? ("--cache-db", $self->{special_protein_cache_db}) : (),
	       $self->{special_protein_cache_dbhost} ? ("--cache-host", $self->{special_protein_cache_dbhost}) : (),
	       $self->{special_protein_cache_dbuser} ? ("--cache-user", $self->{special_protein_cache_dbuser}) : (),
	       $self->{special_protein_cache_dbpass} ? ("--cache-pass", $self->{special_protein_cache_dbpass}) : ());

    $ctx->stderr->log_cmd(@cmd);
    my $ok = run(\@cmd, '>', $out, $ctx->stderr->redirect);
    close($out);
    $results = [];
    if ($ok)
    {
	if (open(my $fh, "<", $out))
	{
	    my $l = <$fh>;
	    while (defined($l = <$fh>))
	    {
		chomp $l;
		push(@$results, [ split(/\t/, $l) ])
	    }
	    close($fh);
	}
    }
    else
    {
	die "Error running rast_compute_specialty_genes ($?): @cmd\n" . $ctx->stderr->text_value;
    }

    #END compute_special_proteins
    my @_bad_returns;
    (ref($results) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"results\" (value was \"$results\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to compute_special_proteins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'compute_special_proteins');
    }
    return($results);
}




=head2 annotate_special_proteins

  $genome_out = $obj->annotate_special_proteins($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_special_proteins
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_special_proteins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_special_proteins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_special_proteins

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $dir = $self->{special_protein_dbdir};
    
    my $prots = File::Temp->new();
    close($prots);
    $genome_in->write_protein_translations_to_file($prots);

    my $out = File::Temp->new();

    my @opts;
    push(@opts, "--parallel", $self->{special_protein_threads});
    if ($self->{special_protein_cache_db})
    {
	push(@opts, "--cache-db", $self->{special_protein_cache_db});
	push(@opts,
	     $self->{special_protein_cache_dbhost} ? ("--cache-host", $self->{special_protein_cache_dbhost}) : (),
	     $self->{special_protein_cache_dbuser} ? ("--cache-user", $self->{special_protein_cache_dbuser}) : (),
	     $self->{special_protein_cache_dbpass} ? ("--cache-pass", $self->{special_protein_cache_dbpass}) : ());
    }
    else
    {
	push(@opts, "--no-cache");
    }

    my @cmd = ('rast_compute_specialty_genes',
	       "--in", $prots,
	       "--db-dir", $dir,
	       @opts);
    
    $ctx->stderr->log_cmd(@cmd);
    my $ok = run(\@cmd, '>', $out, $ctx->stderr->redirect);
    close($out);
    if ($ok)
    {
	if (open(my $fh, "<", $out))
	{
	    my $l = <$fh>;
	    while (defined($l = <$fh>))
	    {
		chomp $l;
		my($qid, $db, $sid, $qcov, $scov, $iden, $pvalue) = split(/\t/, $l);

		my $f = $genome_in->find_feature($qid);
		if (ref($f))
		{
		    push(@{$f->{similarity_associations}}, [$db, $sid, $qcov, $scov, $iden, $pvalue]);
		}
	    }
	    close($fh);
	}
    }
    else
    {
	die "Error running rast_compute_specialty_genes ($?): @cmd\n" . $ctx->stderr->text_value;
    }

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END annotate_special_proteins
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_special_proteins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_special_proteins');
    }
    return($genome_out);
}




=head2 annotate_families_figfam_v1

  $genome_out = $obj->annotate_families_figfam_v1($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_families_figfam_v1
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_families_figfam_v1:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_families_figfam_v1');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_families_figfam_v1

    #
    # Use the patric_call_proteins code from SEED to call figfamv1 families.
    # The path to this tool needs to be configured in the deployment configuration using
    # key patric_call_proteins_path.
    #

    my $remote = $self->{patric_call_proteins_remote_host};
    my $remote_user = $self->{patric_call_proteins_remote_user};
    my $exe = $self->{patric_call_proteins_path};

    if (!$remote && ($exe eq '' || ! -x $exe))
    {
	warn "PATRIC protein caller path not defined or invalid";
	$genome_out = $genome_in;
	goto do_return;
    }

    my $ff = $self->{patric_call_proteins_ff_path};
    my $md5_to_fam = $self->{patric_call_proteins_md5_to_fam_path};

    if (!$remote)
    {
	-d $ff or die "PATRIC protein caller figfam path not found";
	-f $md5_to_fam or die "PATRIC protein caller md5_to_fam path not found";
    }

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $prots = File::Temp->new();
    close($prots);
    $genome_in->write_protein_translations_to_file($prots);

    my $out = File::Temp->new();

    my(@cmd, $ok);
    if ($remote)
    {
	my $key = $self->{patric_call_proteins_remote_key};
	my @kopt = defined($key) ? ("-i", $key) : ();
	my @uopt = defined($remote_user) ? ("-l", $remote_user) : ();
	
	@cmd = ("ssh", @uopt, @kopt, $remote, "$exe -ff $ff --md5-to-fam $md5_to_fam -");
	print Dumper(\@cmd, $prots, $out);
	$ok = run(\@cmd, "<", "$prots", ">", $out, $ctx->stderr->redirect);
	close($out);
    }
    else
    {
	@cmd = ($exe, "-ff", $ff, "--md5-to-fam", $md5_to_fam, $prots);
	$ok = run(\@cmd, ">", $out, $ctx->stderr->redirect);
	close($out);
    }
    if (!$ok)
    {
	die "Error $? running @cmd\n" . $ctx->stderr->text_value;
    }

    open(my $fh, "<", $out) or die "Cannot open output temp $out: $!";

    while (<$fh>)
    {
	chomp;
	my($peg, $kfam, $kfun, @kscore) = split(/\t/);
	next unless $kfam;
	my $f = $genome_in->find_feature($peg);
	if (ref($f))
	{
	    push(@{$f->{family_assignments}}, [ 'FIGFAM', $kfam, $kfun ]);
	}
    }

    close($fh);

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

do_return:
    #END annotate_families_figfam_v1
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_families_figfam_v1:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_families_figfam_v1');
    }
    return($genome_out);
}




=head2 annotate_families_patric

  $genome_out = $obj->annotate_families_patric($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_families_patric
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_families_patric:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_families_patric');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_families_patric

    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @cmd = ("place_proteins_into_pattyfams", "--output", $tmp_out,
	       $self->{patric_annotate_families_kmers},
	       $self->{patric_annotate_families_url},
	       $tmp_in);
    
    my $rc = system(@cmd);
    if ($rc != 0)
    {
	die "error calling patric families: $rc\non command @cmd";
    }

    $genome_out = $coder->decode(scalar read_file("" . $tmp_out));

    #END annotate_families_patric
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_families_patric:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_families_patric');
    }
    return($genome_out);
}




=head2 annotate_null_to_hypothetical

  $genome_out = $obj->annotate_null_to_hypothetical($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_null_to_hypothetical
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_null_to_hypothetical:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_null_to_hypothetical');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_null_to_hypothetical

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $event = {
	tool_name => "annotate_null_to_hypothetical",
	execution_time => scalar gettimeofday,
	hostname => $self->{hostname},
    };
    my $event_id = $genome_in->add_analysis_event($event);

    for my $feature ($genome_in->features)
    {
	if ($feature->{function} eq '')
	{
	    $genome_in->update_function($ctx->user_id, $feature->{id}, "hypothetical protein", $event_id);
	}
    }

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END annotate_null_to_hypothetical
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_null_to_hypothetical:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_null_to_hypothetical');
    }
    return($genome_out);
}




=head2 annotate_strain_type_MLST

  $genome_out = $obj->annotate_strain_type_MLST($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_strain_type_MLST
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_strain_type_MLST:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_strain_type_MLST');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN annotate_strain_type_MLST

    $genome_in = GenomeTypeObject->initialize($genome_in);

    if (!defined($genome_in->{scientific_name}))
    {
	warn "No scientific name defined in annotate_strain_type_MLST";
    }
    else
    {
	my $contigs = $genome_in->extract_contig_sequences_to_temp_file();
	my $tmp_out = File::Temp->new();
	
	my $event = {
	    tool_name => "assign_st_to_genome",
	    execution_time => scalar gettimeofday,
	    hostname => $self->{hostname},
	};
	my $event_id = $genome_in->add_analysis_event($event);
	
	my @cmd = ("assign_st_to_genome",
		   "-s", 1, 
		   "-m", $self->{patric_mlst_dbdir},
		   "--org", $genome_in->{scientific_name});
	$ctx->stderr->log_cmd(@cmd);
	my $ok = run(\@cmd,
		     "<", $contigs,
		     ">", $tmp_out,
		     $ctx->stderr->redirect);
	
	close($tmp_out);
	unlink($contigs);
	
	if (open(my $fh, "<", $tmp_out))
	{
	    while (<$fh>)
	    {
		chomp;
		my($org, $db, $sids, $type, $coords, $loci, $profile, $tag) = split(/\t/);
		push(@{$genome_in->{typing}},
		 {
		     typing_method => 'MLST',
		     database => $db,
		     tag => $tag,
		     event_id => $event_id,
		 });
	    }
	}
    }	
    $genome_out = $genome_in->prepare_for_return();
    
    #END annotate_strain_type_MLST
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_strain_type_MLST:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_strain_type_MLST');
    }
    return($genome_out);
}




=head2 compute_cdd

  $return = $obj->compute_cdd($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$return is a reference to a list where each element is a cdd_hit
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

cdd_hit is a reference to a list containing 16 items:
	0: (protein_id) a string
	1: (domain_id) a string
	2: (identity) a float
	3: (alignment_len) an int
	4: (mismatches) an int
	5: (gap_openings) an int
	6: (protein_start) an int
	7: (protein_end) an int
	8: (domain_start) an int
	9: (domain_end) an int
	10: (e_value) a float
	11: (bit_score) a float
	12: (accession) a string
	13: (short_name) a string
	14: (description) a string
	15: (pssm_length) an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$return is a reference to a list where each element is a cdd_hit
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

cdd_hit is a reference to a list containing 16 items:
	0: (protein_id) a string
	1: (domain_id) a string
	2: (identity) a float
	3: (alignment_len) an int
	4: (mismatches) an int
	5: (gap_openings) an int
	6: (protein_start) an int
	7: (protein_end) an int
	8: (domain_start) an int
	9: (domain_end) an int
	10: (e_value) a float
	11: (bit_score) a float
	12: (accession) a string
	13: (short_name) a string
	14: (description) a string
	15: (pssm_length) an int


=end text



=item Description



=back

=cut

sub compute_cdd
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to compute_cdd:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'compute_cdd');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN compute_cdd
    #END compute_cdd
    my @_bad_returns;
    (ref($return) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to compute_cdd:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'compute_cdd');
    }
    return($return);
}




=head2 annotate_proteins

  $return = $obj->annotate_proteins($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub annotate_proteins
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to annotate_proteins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN annotate_proteins
    $return = $self->assign_functions_to_CDSs($genomeTO);
    #END annotate_proteins
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to annotate_proteins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'annotate_proteins');
    }
    return($return);
}




=head2 estimate_crude_phylogenetic_position_kmer

  $position_estimate = $obj->estimate_crude_phylogenetic_position_kmer($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$position_estimate is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$position_estimate is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Determine close genomes.

=back

=cut

sub estimate_crude_phylogenetic_position_kmer
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to estimate_crude_phylogenetic_position_kmer:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'estimate_crude_phylogenetic_position_kmer');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($position_estimate);
    #BEGIN estimate_crude_phylogenetic_position_kmer
    die "Not implemented";
    #END estimate_crude_phylogenetic_position_kmer
    my @_bad_returns;
    (!ref($position_estimate)) or push(@_bad_returns, "Invalid type for return variable \"position_estimate\" (value was \"$position_estimate\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to estimate_crude_phylogenetic_position_kmer:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'estimate_crude_phylogenetic_position_kmer');
    }
    return($position_estimate);
}




=head2 find_close_neighbors

  $return = $obj->find_close_neighbors($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub find_close_neighbors
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to find_close_neighbors:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'find_close_neighbors');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN find_close_neighbors

    my $kmer_service = Bio::KBase::KmerAnnotationByFigfam::Client->new($self->{kmer_service_url});

    if (! $kmer_service->can("estimate_closest_genomes"))
    {
	die "Kmer service does not support estimate_closest_genomes";
    }

    my @prots = map { [$_->{id}, $_->{function}, $_->{protein_translation}] }
    		grep { $_->{protein_translation} }
    		@{$genomeTO->{features}};
    
    my $dataset_name = $kmer_service->get_default_dataset_name();
    my $close = $kmer_service->estimate_closest_genomes(\@prots, undef);

    my $clist = $genomeTO->{close_genomes};
    if (!ref($clist))
    {
	$clist = [];
	$genomeTO->{close_genomes} = $clist;
    }

    push @$clist, { genome_id => $_->[0],
		    closeness_measure => $_->[1] * 1,
		    genome_name => $_->[2],
		    analysis_method => "KmerAnnotationByFigfam::estimate_closest_genomes:$dataset_name" } foreach @$close;

    $return = $genomeTO;

    #END find_close_neighbors
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to find_close_neighbors:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'find_close_neighbors');
    }
    return($return);
}




=head2 call_features_strep_suis_repeat

  $return = $obj->call_features_strep_suis_repeat($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Interface to Strep repeats and "boxes" tools

=back

=cut

sub call_features_strep_suis_repeat
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_strep_suis_repeat:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_strep_suis_repeat');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_strep_suis_repeat

    $return = $self->_call_using_strep_repeats($genomeTO, "suis_repeat_annotation");

    #END call_features_strep_suis_repeat
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_strep_suis_repeat:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_strep_suis_repeat');
    }
    return($return);
}




=head2 call_features_strep_pneumo_repeat

  $return = $obj->call_features_strep_pneumo_repeat($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_strep_pneumo_repeat
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_strep_pneumo_repeat:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_strep_pneumo_repeat');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN call_features_strep_pneumo_repeat

    $return = $self->_call_using_strep_repeats($genomeTO, "pneumococcal_repeat_annotation");

    #END call_features_strep_pneumo_repeat
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_strep_pneumo_repeat:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_strep_pneumo_repeat');
    }
    return($return);
}




=head2 call_features_crispr

  $genome_out = $obj->call_features_crispr($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub call_features_crispr
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to call_features_crispr:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_crispr');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN call_features_crispr

    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @cmd = ("rast_call_crisprs", "--input", $tmp_in, "--output", $tmp_out,
	       "--id-prefix", $genome_in->{id});
    my $rc = system(@cmd);
    if ($rc != 0)
    {
	die "error calling rRNAs: $rc\non command @cmd";
    }

    $genome_out = $coder->decode(scalar read_file("" . $tmp_out));

    #END call_features_crispr
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to call_features_crispr:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'call_features_crispr');
    }
    return($genome_out);
}




=head2 update_functions

  $genome_out = $obj->update_functions($genome_in, $functions, $event)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$functions is a reference to a list where each element is a reference to a list containing 2 items:
	0: a feature_id
	1: (function) a string
$event is an analysis_event
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$functions is a reference to a list where each element is a reference to a list containing 2 items:
	0: a feature_id
	1: (function) a string
$event is an analysis_event
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description



=back

=cut

sub update_functions
{
    my $self = shift;
    my($genome_in, $functions, $event) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($functions) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"functions\" (value was \"$functions\")");
    (ref($event) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"event\" (value was \"$event\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to update_functions:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'update_functions');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN update_functions

    $genome_in = GenomeTypeObject->initialize($genome_in);

    #
    # If we don't have a valid event, create one here.
    #
    if (!$event->{tool_name})
    {
	$event = {
	    tool_name => "GenomeAnnotation::update_functions",
	    execution_time => scalar gettimeofday,
	    parameters => [],
	    hostname => $self->{hostname},
	};
    }
    my $event_id = $genome_in->add_analysis_event($event);

    for my $ent (@$functions)
    {
	my($fid, $function) = @$ent;
	$genome_in->update_function($ctx->user_id, $fid, $function, $event_id);
    }

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END update_functions
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to update_functions:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'update_functions');
    }
    return($genome_out);
}




=head2 renumber_features

  $genome_out = $obj->renumber_features($genome_in)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Renumber features such that their identifiers are contiguous along contigs.


=back

=cut

sub renumber_features
{
    my $self = shift;
    my($genome_in) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to renumber_features:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'renumber_features');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN renumber_features

    $genome_in = GenomeTypeObject->initialize($genome_in);

    my $event = {
	tool_name => "GenomeAnnotation::renumber_features",
	execution_time => scalar gettimeofday,
	parameters => [],
	hostname => $self->{hostname},
    };

    my $event_id = $genome_in->add_analysis_event($event);

    $genome_in->renumber_features($ctx->user_id, $event_id);

    $genome_out = $genome_in;
    $genome_out = $genome_out->prepare_for_return();

    #END renumber_features
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to renumber_features:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'renumber_features');
    }
    return($genome_out);
}




=head2 classify_amr

  $return = $obj->classify_amr($genomeTO)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genomeTO is a genomeTO
$return is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Perform AMR classification.

=back

=cut

sub classify_amr
{
    my $self = shift;
    my($genomeTO) = @_;

    my @_bad_arguments;
    (ref($genomeTO) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genomeTO\" (value was \"$genomeTO\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to classify_amr:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_amr');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN classify_amr

    my $genome_in = GenomeTypeObject->initialize($genomeTO);

    my $classifier = AdaboostClassify->new();

    my($genus, $species) = $genome_in->{scientific_name} =~ /^(\S+)\s+(\S+)/;
    my $name = "$genus $species";
    my $tmp = File::Temp->new();
    $tmp->close();
    $genome_in->write_contigs_to_file($tmp);
    my $res = $classifier->classify($name, "$tmp");

    my $idc = IDclient->new($genome_in);
    my $event = {
	tool_name => "AdaboostClassify",
	execution_time => scalar gettimeofday,
	parameters => [],
	hostname => $self->{hostname},
    };

    my $event_id = $genome_in->add_analysis_event($event);

    my $type = 'classifier_predicted_region';

    for my $classification (@$res)
    {
	my @flist;
	for my $f (@{$classification->{features}})
	{
	    my($contig, $start, $stop, $alpha, $round, $classifier, $function) = @$f;

	    my $len = $stop - $start + 1;
	    my $loc = [[$contig, $start, "+", $len]];
	    my $feat = $genome_in->add_feature({
		-id_client 	     => $idc,
		-id_prefix 	     => $genome_in->{id},
		-type 	     => $type,
		-location 	     => $loc,
		-function 	     => $function,
		-annotator => "$classifier classifier",
		-annotation      => "Classification by $classifier classifier with alpha=$alpha round=$round",
		-analysis_event_id 	     => $event_id,
	    });
	    push(@flist, [$feat->{id}, $alpha, $round, $function]);
	}
	my $cobj = {
	    name => $classification->{classifier},
	    comment => $classification->{comment},
	    antibiotics =>  [ split(/,\s+/, $classification->{antibiotic}) ],
	    accuracy => $classification->{accuracy},
	    area_under_roc_curve => $classification->{area_under_roc_curve},
	    f1_score => $classification->{f1_score},
	    cumulative_adaboost_score => $classification->{cumulative_adaboost_score},
	    sources => $classification->{sources},
	    sensitivity => $classification->{sensitivity},
	    event_id => $event_id,
	    features => [@flist],
	};
	push(@{$genome_in->{classifications}}, $cobj);
    }

    for my $f (<$tmp*>)
    {
	unlink($f);
    }

    $return = $genome_in->prepare_for_return();
    
    #END classify_amr
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to classify_amr:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_amr');
    }
    return($return);
}




=head2 export_genome

  $exported_data = $obj->export_genome($genome_in, $format, $feature_types)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$format is a string
$feature_types is a reference to a list where each element is a string
$exported_data is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string


</pre>

=end html

=begin text

$genome_in is a genomeTO
$format is a string
$feature_types is a reference to a list where each element is a string
$exported_data is a string
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string



=end text



=item Description

Export genome typed object to one of the supported output formats:
genbank, embl, or gff.
If feature_types is a non-empty list, limit the output to the given
feature types.

=back

=cut

sub export_genome
{
    my $self = shift;
    my($genome_in, $format, $feature_types) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (!ref($format)) or push(@_bad_arguments, "Invalid type for argument \"format\" (value was \"$format\")");
    (ref($feature_types) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"feature_types\" (value was \"$feature_types\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to export_genome:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'export_genome');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($exported_data);
    #BEGIN export_genome

    my $coder = _get_coder();
    my $tmp_in = File::Temp->new();
    write_file($tmp_in, $coder->encode($genome_in));

    close($tmp_in);
    my $tmp_out = File::Temp->new();
    close($tmp_out);

    my @type_flag = map { ("--feature-type", $_) } @$feature_types;

    if ($self->{special_protein_lookup_db})
    {
	push(@type_flag, "--specialty-gene-lookup-db", $self->{special_protein_lookup_db});
    }

    my @cmd = ("rast_export_genome", @type_flag, "--input", $tmp_in, "--output", $tmp_out, $format);

    my $rc = system(@cmd);
    if ($rc != 0)
    {
	die "error exporting genome: $rc\non command @cmd";
    }

    $exported_data = scalar read_file("" . $tmp_out);

    #END export_genome
    my @_bad_returns;
    (!ref($exported_data)) or push(@_bad_returns, "Invalid type for return variable \"exported_data\" (value was \"$exported_data\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to export_genome:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'export_genome');
    }
    return($exported_data);
}




=head2 enumerate_classifiers

  $return = $obj->enumerate_classifiers()

=over 4

=item Parameter and return types

=begin html

<pre>
$return is a reference to a list where each element is a string

</pre>

=end html

=begin text

$return is a reference to a list where each element is a string


=end text



=item Description

Enumerate the available classifiers. Returns the list of identifiers for
the classifiers.

=back

=cut

sub enumerate_classifiers
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN enumerate_classifiers

    $return = [];
    my $dir = $self->{kmer_classifier_data_directory};
    for my $ent (<$dir/*/groups>)
    {
	print STDERR "Try $ent\n";
	my($name) = $ent =~ m,([^/]+)/groups$,;
	push(@$return, $name);
    }
    #END enumerate_classifiers
    my @_bad_returns;
    (ref($return) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to enumerate_classifiers:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'enumerate_classifiers');
    }
    return($return);
}




=head2 query_classifier_groups

  $return = $obj->query_classifier_groups($classifier)

=over 4

=item Parameter and return types

=begin html

<pre>
$classifier is a string
$return is a reference to a hash where the key is a string and the value is a reference to a list where each element is a genome_id
genome_id is a string

</pre>

=end html

=begin text

$classifier is a string
$return is a reference to a hash where the key is a string and the value is a reference to a list where each element is a genome_id
genome_id is a string


=end text



=item Description

Query the groups included in the given classifier. This is a
mapping from the group name to the list of genome IDs included
in the group. Note that these are genome IDs native to the
system that created the classifier; currently these are
SEED genome IDs that may be translated using the
source IDs on the Genome entity.

=back

=cut

sub query_classifier_groups
{
    my $self = shift;
    my($classifier) = @_;

    my @_bad_arguments;
    (!ref($classifier)) or push(@_bad_arguments, "Invalid type for argument \"classifier\" (value was \"$classifier\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to query_classifier_groups:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'query_classifier_groups');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN query_classifier_groups

    my $cobj = KmerClassifier->new("$self->{kmer_classifier_data_directory}/$classifier");
    $return = $cobj->group_membership_hash();

    #END query_classifier_groups
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to query_classifier_groups:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'query_classifier_groups');
    }
    return($return);
}




=head2 query_classifier_taxonomies

  $return = $obj->query_classifier_taxonomies($classifier)

=over 4

=item Parameter and return types

=begin html

<pre>
$classifier is a string
$return is a reference to a hash where the key is a string and the value is a string

</pre>

=end html

=begin text

$classifier is a string
$return is a reference to a hash where the key is a string and the value is a string


=end text



=item Description

Query the taxonomy strings that this classifier maps.

=back

=cut

sub query_classifier_taxonomies
{
    my $self = shift;
    my($classifier) = @_;

    my @_bad_arguments;
    (!ref($classifier)) or push(@_bad_arguments, "Invalid type for argument \"classifier\" (value was \"$classifier\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to query_classifier_taxonomies:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'query_classifier_taxonomies');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN query_classifier_taxonomies
    #END query_classifier_taxonomies
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to query_classifier_taxonomies:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'query_classifier_taxonomies');
    }
    return($return);
}




=head2 classify_into_bins

  $return = $obj->classify_into_bins($classifier, $dna_input)

=over 4

=item Parameter and return types

=begin html

<pre>
$classifier is a string
$dna_input is a reference to a list where each element is a reference to a list containing 2 items:
	0: (id) a string
	1: (dna_data) a string
$return is a reference to a hash where the key is a string and the value is an int

</pre>

=end html

=begin text

$classifier is a string
$dna_input is a reference to a list where each element is a reference to a list containing 2 items:
	0: (id) a string
	1: (dna_data) a string
$return is a reference to a hash where the key is a string and the value is an int


=end text



=item Description

Classify a dataset, returning only the binned output.

=back

=cut

sub classify_into_bins
{
    my $self = shift;
    my($classifier, $dna_input) = @_;

    my @_bad_arguments;
    (!ref($classifier)) or push(@_bad_arguments, "Invalid type for argument \"classifier\" (value was \"$classifier\")");
    (ref($dna_input) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"dna_input\" (value was \"$dna_input\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to classify_into_bins:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_into_bins');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN classify_into_bins

    my $cobj = KmerClassifier->new("$self->{kmer_classifier_data_directory}/$classifier");

    my $tmp = File::Temp->new;
    close($tmp);
    open(TMP, ">", $tmp) or die "cannot write tempfile $tmp: $!";
    for my $ent (@$dna_input)
    {
	gjoseqlib::print_alignment_as_fasta(\*TMP, [$ent->[0], undef, $ent->[1]]);
    }
    close(TMP);
    my($bins, $missed) = $cobj->classify("" . $tmp);

    $return = $bins;

    #END classify_into_bins
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to classify_into_bins:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_into_bins');
    }
    return($return);
}




=head2 classify_full

  $return_1, $raw_output, $unassigned = $obj->classify_full($classifier, $dna_input)

=over 4

=item Parameter and return types

=begin html

<pre>
$classifier is a string
$dna_input is a reference to a list where each element is a reference to a list containing 2 items:
	0: (id) a string
	1: (dna_data) a string
$return_1 is a reference to a hash where the key is a string and the value is an int
$raw_output is a string
$unassigned is a reference to a list where each element is a string

</pre>

=end html

=begin text

$classifier is a string
$dna_input is a reference to a list where each element is a reference to a list containing 2 items:
	0: (id) a string
	1: (dna_data) a string
$return_1 is a reference to a hash where the key is a string and the value is an int
$raw_output is a string
$unassigned is a reference to a list where each element is a string


=end text



=item Description

Classify a dataset, returning the binned output along with the raw assignments and the list of
sequences that were not assigned.

=back

=cut

sub classify_full
{
    my $self = shift;
    my($classifier, $dna_input) = @_;

    my @_bad_arguments;
    (!ref($classifier)) or push(@_bad_arguments, "Invalid type for argument \"classifier\" (value was \"$classifier\")");
    (ref($dna_input) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"dna_input\" (value was \"$dna_input\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to classify_full:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_full');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return_1, $raw_output, $unassigned);
    #BEGIN classify_full

    my $cobj = KmerClassifier->new("$self->{kmer_classifier_data_directory}/$classifier");

    my $tmp = File::Temp->new;
    close($tmp);
    open(TMP, ">", $tmp) or die "cannot write tempfile $tmp: $!";

    my $raw_tmp = File::Temp->new();
    for my $ent (@$dna_input)
    {
	gjoseqlib::print_alignment_as_fasta(\*TMP, [$ent->[0], undef, $ent->[1]]);
    }
    close(TMP);
    my($bins, $missed) = $cobj->classify("" . $tmp, $raw_tmp);
    close($raw_tmp);

    $return_1 = $bins;
    $unassigned = $missed;
    $raw_output = read_file("" . $raw_tmp);

    #END classify_full
    my @_bad_returns;
    (ref($return_1) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return_1\" (value was \"$return_1\")");
    (!ref($raw_output)) or push(@_bad_returns, "Invalid type for return variable \"raw_output\" (value was \"$raw_output\")");
    (ref($unassigned) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"unassigned\" (value was \"$unassigned\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to classify_full:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'classify_full');
    }
    return($return_1, $raw_output, $unassigned);
}




=head2 default_workflow

  $return = $obj->default_workflow()

=over 4

=item Parameter and return types

=begin html

<pre>
$return is a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$return is a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub default_workflow
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN default_workflow

    my @stages = (
	      { name => 'call_features_rRNA_SEED' },
	      { name => 'call_features_tRNA_trnascan' },
	      { name => 'call_features_repeat_region_SEED',
		    repeat_region_SEED_parameters => { } },
	      { name => 'call_selenoproteins', failure_is_not_fatal => 1 },
	      { name => 'call_pyrrolysoproteins', failure_is_not_fatal => 1 },
	      { name => 'call_features_strep_suis_repeat',
		condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
	      { name => 'call_features_strep_pneumo_repeat',
		condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
	      { name => 'call_features_crispr', failure_is_not_fatal => 1 },
	      { name => 'call_features_CDS_prodigal' },
	      { name => 'call_features_CDS_glimmer3', failure_is_not_fatal => 1, glimmer3_parameters => {} },
	      { name => 'annotate_proteins_kmer_v2', kmer_v2_parameters => {} },
	      { name => 'annotate_proteins_kmer_v1', kmer_v1_parameters => { annotate_hypothetical_only => 1 } },
	      { name => 'annotate_proteins_phage', phage_parameters => { annotate_hypothetical_only => 1 } },
	      { name => 'annotate_proteins_similarity', similarity_parameters => { annotate_hypothetical_only => 1 } },
	      { name => 'propagate_genbank_feature_metadata', propagate_genbank_feature_metadata_parameters => {} },
	      { name => 'resolve_overlapping_features', resolve_overlapping_features_parameters => {} },
	      { name => 'classify_amr', failure_is_not_fatal => 1 },
	      { name => 'renumber_features' },
              { name => 'annotate_special_proteins', failure_is_not_fatal => 1 },
	      { name => 'annotate_families_figfam_v1', failure_is_not_fatal => 1 },
	      { name => 'annotate_families_patric', failure_is_not_fatal => 1 },
	      { name => 'annotate_null_to_hypothetical' },
	      { name => 'find_close_neighbors', failure_is_not_fatal => 1 },
              { name => 'annotate_strain_type_MLST', failure_is_not_fatal => 1  },
	      # { name => 'call_features_prophage_phispy' },
		 );
    $return = { stages => \@stages };
    
    #END default_workflow
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to default_workflow:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'default_workflow');
    }
    return($return);
}




=head2 enumerate_workflows

  $workflows = $obj->enumerate_workflows()

=over 4

=item Parameter and return types

=begin html

<pre>
$workflows is a reference to a list where each element is a reference to a list containing 2 items:
	0: (workflow_id) a string
	1: (wf) a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$workflows is a reference to a list where each element is a reference to a list containing 2 items:
	0: (workflow_id) a string
	1: (wf) a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description

Enumerate the loaded workflows. We always have a workflow named "default"; a
particular deployment of the genome annotation service may include additional workflows.

=back

=cut

sub enumerate_workflows
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($workflows);
    #BEGIN enumerate_workflows

    $workflows = [];
    
    my $def = $self->default_workflow();
    push(@$workflows, ["default", $def]);

    #
    # If we have a workflow directory configured, read from that.
    #
    my $dir = $self->{workflow_dir};
    if ($dir && opendir(my $dh, $dir))
    {
	my $coder = _get_coder();

	for my $name (sort { $a cmp $b } grep { -f "$dir/$_" && $_ ne 'default' } readdir($dh))
	{
	    my $data;
	    eval {
		my $txt = read_file("$dir/$name");
		$data = $coder->decode($txt);
	    };
	    if ($@)
	    {
		warn "error reading $dir/$name: $@";
	    }
	    else
	    {
		push(@$workflows, [$name, $data]);
	    }
	}
	closedir($dh);
    }
    
    #END enumerate_workflows
    my @_bad_returns;
    (ref($workflows) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"workflows\" (value was \"$workflows\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to enumerate_workflows:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'enumerate_workflows');
    }
    return($workflows);
}




=head2 complete_workflow_template

  $return = $obj->complete_workflow_template()

=over 4

=item Parameter and return types

=begin html

<pre>
$return is a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$return is a workflow
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description

Return a workflow that includes all available stages. Not meant
(necessarily) for actual execution, but as a comprehensive list
of parts for users to use in assembling their own workflows.

=back

=cut

sub complete_workflow_template
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($return);
    #BEGIN complete_workflow_template

    my @stages = (
	      { name => 'call_features_rRNA_SEED' },
	      { name => 'call_features_tRNA_trnascan' },
	      { name => 'call_features_repeat_region_SEED',
		    repeat_region_SEED_parameters => { } },
	      { name => 'call_selenoproteins' },
	      { name => 'call_pyrrolysoproteins' },
	      { name => 'call_features_strep_suis_repeat',
		condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
	      { name => 'call_features_strep_pneumo_repeat',
		condition => '$genome->{scientific_name} =~ /^Streptococcus\s/' },
	      { name => 'call_features_crispr', failure_is_not_fatal => 1 },
	      { name => 'call_features_CDS_glimmer3', glimmer3_parameters => { min_training_len => 2000 } },
	      { name => 'call_features_CDS_prodigal' },
	      { name => 'call_features_CDS_genemark' },
	      { name => 'annotate_proteins_kmer_v2', kmer_v2_parameters => { annotate_hypothetical_only => 0 } },
	      { name => 'annotate_proteins_kmer_v1', kmer_v1_parameters => { annotate_hypothetical_only => 1 } },
	      { name => 'annotate_proteins_similarity', similarity_parameters => { annotate_hypothetical_only => 1 } },
	      { name => 'annotate_null_to_hypothetical' },
	      { name => 'propagate_genbank_feature_metadata', propagate_genbank_feature_metadata_parameters => {} },
	      { name => 'resolve_overlapping_features', resolve_overlapping_features_parameters => {} },
	      { name => 'renumber_features' },
	      { name => 'find_close_neighbors', failure_is_not_fatal => 1 },
	      { name => 'call_features_prophage_phispy' },
		 );
    $return = { stages => \@stages };

    #END complete_workflow_template
    my @_bad_returns;
    (ref($return) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"return\" (value was \"$return\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to complete_workflow_template:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'complete_workflow_template');
    }
    return($return);
}




=head2 run_pipeline

  $genome_out = $obj->run_pipeline($genome_in, $workflow)

=over 4

=item Parameter and return types

=begin html

<pre>
$genome_in is a genomeTO
$workflow is a workflow
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genome_in is a genomeTO
$workflow is a workflow
$genome_out is a genomeTO
genomeTO is a reference to a hash where the following keys are defined:
	id has a value which is a genome_id
	scientific_name has a value which is a string
	domain has a value which is a string
	genetic_code has a value which is an int
	source has a value which is a string
	source_id has a value which is a string
	taxonomy has a value which is a string
	ncbi_taxonomy_id has a value which is an int
	owner has a value which is a string
	quality has a value which is a genome_quality_measure
	contigs has a value which is a reference to a list where each element is a contig
	contigs_handle has a value which is a Handle
	features has a value which is a reference to a list where each element is a feature
	close_genomes has a value which is a reference to a list where each element is a close_genome
	analysis_events has a value which is a reference to a list where each element is an analysis_event
	typing has a value which is a reference to a list where each element is a strain_type
	classifications has a value which is a reference to a list where each element is a classifier
genome_id is a string
genome_quality_measure is a reference to a hash where the following keys are defined:
	frameshift_error_rate has a value which is a float
	sequence_error_rate has a value which is a float
contig is a reference to a hash where the following keys are defined:
	id has a value which is a contig_id
	dna has a value which is a string
	genetic_code has a value which is an int
	cell_compartment has a value which is a string
	replicon_type has a value which is a string
	replicon_geometry has a value which is a string
	complete has a value which is a bool
	genbank_locus has a value which is a genbank_locus
contig_id is a string
bool is an int
genbank_locus is a reference to a hash where the following keys are defined:
	accession has a value which is a reference to a list where each element is a string
	comment has a value which is a reference to a list where each element is a string
	date has a value which is a string
	dblink has a value which is a reference to a list where each element is a string
	dbsource has a value which is a reference to a list where each element is a string
	definition has a value which is a string
	division has a value which is a string
	geometry has a value which is a string
	gi has a value which is an int
	keywords has a value which is a reference to a list where each element is a string
	locus has a value which is a string
	organism has a value which is a string
	origin has a value which is a string
	references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
	source has a value which is a string
	taxonomy has a value which is a reference to a list where each element is a string
	version has a value which is a reference to a list where each element is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
feature is a reference to a hash where the following keys are defined:
	id has a value which is a feature_id
	location has a value which is a location
	type has a value which is a feature_type
	function has a value which is a string
	function_id has a value which is a string
	protein_translation has a value which is a string
	aliases has a value which is a reference to a list where each element is a string
	alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
	0: (source) a string
	1: (alias) a string

	annotations has a value which is a reference to a list where each element is an annotation
	quality has a value which is a feature_quality_measure
	feature_creation_event has a value which is an analysis_event_id
	family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
	similarity_associations has a value which is a reference to a list where each element is a similarity_association
	proposed_functions has a value which is a reference to a list where each element is a proposed_function
	genbank_type has a value which is a string
	genbank_feature has a value which is a genbank_feature
feature_id is a string
location is a reference to a list where each element is a region_of_dna
region_of_dna is a reference to a list containing 4 items:
	0: a contig_id
	1: (begin) an int
	2: (strand) a string
	3: (length) an int
feature_type is a string
annotation is a reference to a list containing 4 items:
	0: (comment) a string
	1: (annotator) a string
	2: (annotation_time) a float
	3: an analysis_event_id
analysis_event_id is a string
feature_quality_measure is a reference to a hash where the following keys are defined:
	truncated_begin has a value which is a bool
	truncated_end has a value which is a bool
	existence_confidence has a value which is a float
	frameshifted has a value which is a bool
	selenoprotein has a value which is a bool
	pyrrolysylprotein has a value which is a bool
	overlap_rules has a value which is a reference to a list where each element is a string
	existence_priority has a value which is a float
	hit_count has a value which is a float
	weighted_hit_count has a value which is a float
	genemark_score has a value which is a float
protein_family_assignment is a reference to a list containing 4 items:
	0: (db) a string
	1: (id) a string
	2: (function) a string
	3: (db_version) a string
similarity_association is a reference to a list containing 6 items:
	0: (source) a string
	1: (source_id) a string
	2: (query_coverage) a float
	3: (subject_coverage) a float
	4: (identity) a float
	5: (e_value) a float
proposed_function is a reference to a hash where the following keys are defined:
	id has a value which is a string
	function has a value which is a string
	user has a value which is a string
	score has a value which is a float
	event_id has a value which is an analysis_event_id
	timestamp has a value which is an int
genbank_feature is a reference to a hash where the following keys are defined:
	genbank_type has a value which is a string
	genbank_location has a value which is a string
	values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string
close_genome is a reference to a hash where the following keys are defined:
	genome has a value which is a genome_id
	genome_name has a value which is a string
	closeness_measure has a value which is a float
	analysis_method has a value which is a string
analysis_event is a reference to a hash where the following keys are defined:
	id has a value which is an analysis_event_id
	tool_name has a value which is a string
	execution_time has a value which is a float
	parameters has a value which is a reference to a list where each element is a string
	hostname has a value which is a string
strain_type is a reference to a hash where the following keys are defined:
	typing_method has a value which is a string
	database has a value which is a string
	tag has a value which is a string
	event_id has a value which is an analysis_event_id
classifier is a reference to a hash where the following keys are defined:
	name has a value which is a string
	version has a value which is a string
	description has a value which is a string
	comment has a value which is a string
	antibiotics has a value which is a reference to a list where each element is a string
	accuracy has a value which is a float
	area_under_roc_curve has a value which is a float
	f1_score has a value which is a float
	sources has a value which is a string
	cumulative_adaboost_value has a value which is a float
	sensitivity has a value which is a string
	event_id has a value which is an analysis_event_id
	features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
	0: (id) a feature_id
	1: (alpha) a float
	2: (round) an int
	3: (function) a string

workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub run_pipeline
{
    my $self = shift;
    my($genome_in, $workflow) = @_;

    my @_bad_arguments;
    (ref($genome_in) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"genome_in\" (value was \"$genome_in\")");
    (ref($workflow) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"workflow\" (value was \"$workflow\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to run_pipeline:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'run_pipeline');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($genome_out);
    #BEGIN run_pipeline

    my %param_defs = (annotate_proteins_kmer_v1 => 'kmer_v1_parameters',
		      annotate_proteins_kmer_v2 => 'kmer_v2_parameters',
		      annotate_proteins_similarity => 'similarity_parameters',
		      annotate_proteins_phage => 'phage_parameters',
		      call_features_repeat_region_SEED => 'repeat_regions_SEED_parameters',
		      call_features_CDS_glimmer3 => 'glimmer3_parameters',
		      call_features_repeat_region_SEED => 'repeat_region_SEED_parameters',
		      call_features_CDS_SEED_projection => 'SEED_projection_parameters',
		      call_features_ProtoCDS_kmer_v1 => 'kmer_v1_parameters',
		      propagate_genbank_feature_metadata => 'propagate_genbank_feature_metadata_parameters',
		      call_features_ProtoCDS_kmer_v2 => 'kmer_v2_parameters',
		      resolve_overlapping_features => 'resolve_overlapping_features_parameters',
		      propagate_genbank_feature_metadata => 'propagate_genbank_feature_metadata_parameters',
		      );

    my $cur = $genome_in;
    for my $stage (@{$workflow->{stages}})
    {
	my $method = $stage->{name};
	my $condition = $stage->{condition};
	if ($condition)
	{
	    my $safe = Safe->new();
	    my $g = $safe->varglob('genome');
	    $$g = $cur;
	    my $ok = $safe->reval($condition);
	    print STDERR "Condition eval of '$condition' returns $ok\n";
	    if (!$ok)
	    {
		print STDERR "Skipping $method due to condition $condition\n";
		next;
	    }
	}
	my @params;
	if (my $param_def = $param_defs{$method})
	{
	    push(@params, $stage->{$param_def});
	}
	elsif ($method eq 'call_features_rRNA_SEED')
	{
	    # Special case.
	    push(@params, []);
	}
	if ($self->can($method))
	{
	    print STDERR "Call $method with " . Dumper(\@params);
	    print STDERR Dumper($stage);
	    my $out;
	    eval {
		$out = $self->$method($cur, @params);
	    };

	    if ($@)
	    {
		if ($stage->{failure_is_not_fatal})
		{
		    warn "Error invoking method $method: $@\nContinuing because failure_is_not_fatal flag is set";
		    if (ref($cur) && ref($cur) ne 'HASH')
		    {
			$cur = $cur->prepare_for_return;
		    }
		}
		else
		{
		    die "Error invoking method $method: $@";
		}
	    }
	    else
	    {
		print STDERR "Finished\n";
		$cur = $out;
	    }
	}
	else
	{
	    die "Trying to call invalid method $method";
	}
    }

    $genome_out = $cur;

    #END run_pipeline
    my @_bad_returns;
    (ref($genome_out) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"genome_out\" (value was \"$genome_out\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to run_pipeline:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'run_pipeline');
    }
    return($genome_out);
}




=head2 pipeline_batch_start

  $batch_id = $obj->pipeline_batch_start($genomes, $workflow)

=over 4

=item Parameter and return types

=begin html

<pre>
$genomes is a reference to a list where each element is a pipeline_batch_input
$workflow is a workflow
$batch_id is a string
pipeline_batch_input is a reference to a hash where the following keys are defined:
	genome_id has a value which is a string
	data has a value which is a Handle
	filename has a value which is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

$genomes is a reference to a list where each element is a pipeline_batch_input
$workflow is a workflow
$batch_id is a string
pipeline_batch_input is a reference to a hash where the following keys are defined:
	genome_id has a value which is a string
	data has a value which is a Handle
	filename has a value which is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string
workflow is a reference to a hash where the following keys are defined:
	stages has a value which is a reference to a list where each element is a pipeline_stage
pipeline_stage is a reference to a hash where the following keys are defined:
	name has a value which is a string
	condition has a value which is a string
	failure_is_not_fatal has a value which is an int
	repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
	glimmer3_parameters has a value which is a glimmer3_parameters
	kmer_v1_parameters has a value which is a kmer_v1_parameters
	kmer_v2_parameters has a value which is a kmer_v2_parameters
	similarity_parameters has a value which is a similarity_parameters
repeat_region_SEED_parameters is a reference to a hash where the following keys are defined:
	min_identity has a value which is a float
	min_length has a value which is an int
glimmer3_parameters is a reference to a hash where the following keys are defined:
	min_training_len has a value which is an int
kmer_v1_parameters is a reference to a hash where the following keys are defined:
	kmer_size has a value which is an int
	dataset_name has a value which is a string
	return_scores_for_all_proteins has a value which is an int
	score_threshold has a value which is an int
	hit_threshold has a value which is an int
	sequential_hit_threshold has a value which is an int
	detailed has a value which is an int
	min_hits has a value which is an int
	min_size has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
kmer_v2_parameters is a reference to a hash where the following keys are defined:
	min_hits has a value which is an int
	max_gap has a value which is an int
	annotate_hypothetical_only has a value which is an int
similarity_parameters is a reference to a hash where the following keys are defined:
	annotate_hypothetical_only has a value which is an int


=end text



=item Description



=back

=cut

sub pipeline_batch_start
{
    my $self = shift;
    my($genomes, $workflow) = @_;

    my @_bad_arguments;
    (ref($genomes) eq 'ARRAY') or push(@_bad_arguments, "Invalid type for argument \"genomes\" (value was \"$genomes\")");
    (ref($workflow) eq 'HASH') or push(@_bad_arguments, "Invalid type for argument \"workflow\" (value was \"$workflow\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to pipeline_batch_start:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'pipeline_batch_start');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($batch_id);
    #BEGIN pipeline_batch_start

    #
    # Construct an AWE workflow and submit it to our AWE server.
    #
    # The genomes have already been uploaded - the genomes list here is a
    # list of handle objects.
    #
    # The workflow we create will have one task per genome submitted. The data
    # passed to each task are the serialized form of the handle and the
    # workflow description.
    #

    my $json = JSON::XS->new->pretty(1);
    my $shock = Bio::KBase::GenomeAnnotation::Shock->new($self->{shock_server}, $ctx->token);
    my $awe = Bio::KBase::GenomeAnnotation::Awe->new($self->{awe_server}, $ctx->token);
    
    my $job = Bio::KBase::GenomeAnnotation::Awe::JobDescription->new(pipeline => 'rasttk',
								     name => 'rasttk',
								     project => 'rasttk',
								     user => $ctx->user_id,
								     clientgroups => '');
    
    my $i = 0;
    for my $gspec (@$genomes)
    {
	my($genome_id, $handle, $filename) = @$gspec{'genome_id', 'data', 'filename'};
	
	my $txt = $json->encode([$handle, $workflow]);
	my $node = $shock->put_file_data($txt);
	
	my $in_file = Bio::KBase::GenomeAnnotation::Awe::JobFile->new("pipeinput_$i.json", $shock->server, $node);
	my $out_file = Bio::KBase::GenomeAnnotation::Awe::JobFile->new("pipeoutput_$i.json", $shock->server);
	my $stdout_file = Bio::KBase::GenomeAnnotation::Awe::JobFile->new("stdout_$i.txt", $shock->server);
	my $stderr_file = Bio::KBase::GenomeAnnotation::Awe::JobFile->new("stderr_$i.txt", $shock->server);
	$i++;
	
	my $id = $job->add_task("rast_run_pipeline_local",
				"rast_run_pipeline_local",
				join(" ",
				     $in_file->in_name,
				     $out_file->name,
				     $stdout_file->name,
				     $stderr_file->name),
				[],
				[$in_file],
				[$out_file, $stdout_file, $stderr_file],
				undef, undef, $awe,
			        { genome_id => $genome_id, filename => $filename });
    }
    
    $batch_id = $awe->submit($job);
    print STDERR "submitted $batch_id\n";
    
    #END pipeline_batch_start
    my @_bad_returns;
    (!ref($batch_id)) or push(@_bad_returns, "Invalid type for return variable \"batch_id\" (value was \"$batch_id\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to pipeline_batch_start:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'pipeline_batch_start');
    }
    return($batch_id);
}




=head2 pipeline_batch_status

  $status = $obj->pipeline_batch_status($batch_id)

=over 4

=item Parameter and return types

=begin html

<pre>
$batch_id is a string
$status is a pipeline_batch_status
pipeline_batch_status is a reference to a hash where the following keys are defined:
	status has a value which is a string
	submit_date has a value which is a string
	start_date has a value which is a string
	completion_date has a value which is a string
	details has a value which is a reference to a list where each element is a pipeline_batch_status_entry
pipeline_batch_status_entry is a reference to a hash where the following keys are defined:
	genome_id has a value which is a string
	status has a value which is a string
	creation_date has a value which is a string
	start_date has a value which is a string
	completion_date has a value which is a string
	stdout has a value which is a Handle
	stderr has a value which is a Handle
	output has a value which is a Handle
	filename has a value which is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string

</pre>

=end html

=begin text

$batch_id is a string
$status is a pipeline_batch_status
pipeline_batch_status is a reference to a hash where the following keys are defined:
	status has a value which is a string
	submit_date has a value which is a string
	start_date has a value which is a string
	completion_date has a value which is a string
	details has a value which is a reference to a list where each element is a pipeline_batch_status_entry
pipeline_batch_status_entry is a reference to a hash where the following keys are defined:
	genome_id has a value which is a string
	status has a value which is a string
	creation_date has a value which is a string
	start_date has a value which is a string
	completion_date has a value which is a string
	stdout has a value which is a Handle
	stderr has a value which is a Handle
	output has a value which is a Handle
	filename has a value which is a string
Handle is a reference to a hash where the following keys are defined:
	file_name has a value which is a string
	id has a value which is a string
	type has a value which is a string
	url has a value which is a string
	remote_md5 has a value which is a string
	remote_sha1 has a value which is a string


=end text



=item Description



=back

=cut

sub pipeline_batch_status
{
    my $self = shift;
    my($batch_id) = @_;

    my @_bad_arguments;
    (!ref($batch_id)) or push(@_bad_arguments, "Invalid type for argument \"batch_id\" (value was \"$batch_id\")");
    if (@_bad_arguments) {
	my $msg = "Invalid arguments passed to pipeline_batch_status:\n" . join("", map { "\t$_\n" } @_bad_arguments);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'pipeline_batch_status');
    }

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($status);
    #BEGIN pipeline_batch_status

    my $awe = Bio::KBase::GenomeAnnotation::Awe->new($self->{awe_server}, $ctx->token);
    
    my $job = $awe->job($batch_id);

    #
    # We could check to see if the job's user matches the authenticated user here. I
    # don't think it's necessary at this point.
    #

    #
    # Loop over the tasks collecting the status / file information to be formatted and returned.
    #

    my $tasks = $job->{tasks};

    my $details = [];

    for my $task (@$tasks)
    {
	my $tinfo = {
	    genome_id => $task->{userattr}->{genome_id},
	    filename => $task->{userattr}->{filename},
	    status => $task->{state},
	    creation_date => $task->{createddate},
	    start_date => $task->{starteddate},
	    completion_date => $task->{completeddate},
	};

	my $outfiles = $task->{outputs};
	#
	# Support both old and new format for file list by constructing
	# a new format list from the old.
	#
	if (ref($outfiles) eq 'HASH')
	{
	    my $new = [];
	    while (my($file, $hash) = each %$outfiles)
	    {
		my $nhash = { filename => $file, %$hash };
		push(@$new, $nhash);
	    }
	    $outfiles = $new;
	}
	    
	for my $f (@$outfiles)
	{
	    my $key;
	    my $filename = $f->{filename};
	    if ($filename =~ /^stderr/) { $key = 'stderr'; }
	    elsif ($filename =~ /^stdout/ ) { $key = 'stdout'; }
	    elsif ($filename =~ /^pipeoutput/) { $key = 'output'; }
	    else { next; }
	
	    if ($f->{url} ne '')
	    {
		$tinfo->{$key} = {
		    file_name => $filename,
		    id => $f->{node},
		    type => 'shock',
		    url => $f->{host},
		    remote_md5 => '',
		    remote_sha1 => '',
		};
	    }
	}
	push (@$details, $tinfo);
    }

    $status = {
	status => $job->{info}->{state},
	submit_date => $job->{info}->{submittime},
	start_date => $job->{info}->{startedtime},
	completion_date => $job->{info}->{completedtime},
	details => $details,
    };
    
    #END pipeline_batch_status
    my @_bad_returns;
    (ref($status) eq 'HASH') or push(@_bad_returns, "Invalid type for return variable \"status\" (value was \"$status\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to pipeline_batch_status:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'pipeline_batch_status');
    }
    return($status);
}




=head2 pipeline_batch_enumerate_batches

  $batches = $obj->pipeline_batch_enumerate_batches()

=over 4

=item Parameter and return types

=begin html

<pre>
$batches is a reference to a list where each element is a reference to a list containing 2 items:
	0: (batch_id) a string
	1: (submit_time) a string

</pre>

=end html

=begin text

$batches is a reference to a list where each element is a reference to a list containing 2 items:
	0: (batch_id) a string
	1: (submit_time) a string


=end text



=item Description



=back

=cut

sub pipeline_batch_enumerate_batches
{
    my $self = shift;

    my $ctx = $Bio::KBase::GenomeAnnotation::Service::CallContext;
    my($batches);
    #BEGIN pipeline_batch_enumerate_batches

    #
    # Call AWE to find all tasks with project=rasttk and user=logged-in user
    #
    
    my $awe = Bio::KBase::GenomeAnnotation::Awe->new($self->{awe_server}, $ctx->token);
    my($list, $error) = $awe->enumerate_jobs({ "info.project" => 'rasttk', "info.user" => $ctx->user_id });

    $batches = [];
    if (ref($list))
    {
	$batches = [ map { [ $_->{id}, $_->{info}->{submittime} ] } @$list ];
    }
    else
    {
	die "pipeline_batch_enumerate_batches task query error: $error";
    }

    #END pipeline_batch_enumerate_batches
    my @_bad_returns;
    (ref($batches) eq 'ARRAY') or push(@_bad_returns, "Invalid type for return variable \"batches\" (value was \"$batches\")");
    if (@_bad_returns) {
	my $msg = "Invalid returns passed to pipeline_batch_enumerate_batches:\n" . join("", map { "\t$_\n" } @_bad_returns);
	Bio::KBase::Exceptions::ArgumentValidationError->throw(error => $msg,
							       method_name => 'pipeline_batch_enumerate_batches');
    }
    return($batches);
}




=head2 version 

  $return = $obj->version()

=over 4

=item Parameter and return types

=begin html

<pre>
$return is a string
</pre>

=end html

=begin text

$return is a string

=end text

=item Description

Return the module version. This is a Semantic Versioning number.

=back

=cut

sub version {
    return $VERSION;
}

=head1 TYPES



=head2 Handle

=over 4



=item Description

* This is a handle service handle object, used for by-reference
* passing of data files.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
file_name has a value which is a string
id has a value which is a string
type has a value which is a string
url has a value which is a string
remote_md5 has a value which is a string
remote_sha1 has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
file_name has a value which is a string
id has a value which is a string
type has a value which is a string
url has a value which is a string
remote_md5 has a value which is a string
remote_sha1 has a value which is a string


=end text

=back



=head2 bool

=over 4



=item Definition

=begin html

<pre>
an int
</pre>

=end html

=begin text

an int

=end text

=back



=head2 md5

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 md5s

=over 4



=item Definition

=begin html

<pre>
a reference to a list where each element is a md5
</pre>

=end html

=begin text

a reference to a list where each element is a md5

=end text

=back



=head2 genome_id

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 feature_id

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 contig_id

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 feature_type

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 region_of_dna

=over 4



=item Description

A region of DNA is maintained as a tuple of four components:

                the contig
                the beginning position (from 1)
                the strand
                the length

           We often speak of "a region".  By "location", we mean a sequence
           of regions from the same genome (perhaps from distinct contigs).

           Strand is either '+' or '-'.


=item Definition

=begin html

<pre>
a reference to a list containing 4 items:
0: a contig_id
1: (begin) an int
2: (strand) a string
3: (length) an int

</pre>

=end html

=begin text

a reference to a list containing 4 items:
0: a contig_id
1: (begin) an int
2: (strand) a string
3: (length) an int


=end text

=back



=head2 location

=over 4



=item Description

a "location" refers to a sequence of regions


=item Definition

=begin html

<pre>
a reference to a list where each element is a region_of_dna
</pre>

=end html

=begin text

a reference to a list where each element is a region_of_dna

=end text

=back



=head2 analysis_event_id

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 analysis_event

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is an analysis_event_id
tool_name has a value which is a string
execution_time has a value which is a float
parameters has a value which is a reference to a list where each element is a string
hostname has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is an analysis_event_id
tool_name has a value which is a string
execution_time has a value which is a float
parameters has a value which is a reference to a list where each element is a string
hostname has a value which is a string


=end text

=back



=head2 annotation

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 4 items:
0: (comment) a string
1: (annotator) a string
2: (annotation_time) a float
3: an analysis_event_id

</pre>

=end html

=begin text

a reference to a list containing 4 items:
0: (comment) a string
1: (annotator) a string
2: (annotation_time) a float
3: an analysis_event_id


=end text

=back



=head2 feature_quality_measure

=over 4



=item Description

* The numeric priority of this feature's right to exist. Specialty
* tools will give the features they create a high priority; more generic
* tools will give their features a lower priority. The overlap removal procedure
* will use this priority to determine which of a set of overlapping features
* should be removed.
*
* The intent is that a change of 1 in the priority value represents a factor of 2 in
* preference.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
truncated_begin has a value which is a bool
truncated_end has a value which is a bool
existence_confidence has a value which is a float
frameshifted has a value which is a bool
selenoprotein has a value which is a bool
pyrrolysylprotein has a value which is a bool
overlap_rules has a value which is a reference to a list where each element is a string
existence_priority has a value which is a float
hit_count has a value which is a float
weighted_hit_count has a value which is a float
genemark_score has a value which is a float

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
truncated_begin has a value which is a bool
truncated_end has a value which is a bool
existence_confidence has a value which is a float
frameshifted has a value which is a bool
selenoprotein has a value which is a bool
pyrrolysylprotein has a value which is a bool
overlap_rules has a value which is a reference to a list where each element is a string
existence_priority has a value which is a float
hit_count has a value which is a float
weighted_hit_count has a value which is a float
genemark_score has a value which is a float


=end text

=back



=head2 protein_family_assignment

=over 4



=item Description

* A protein family assignment notes the assignment of the given feature
* to a protein family. db is the name of the protein family database
* (e.g. FIGfam, GPF for GlobalPatricFam, LPF for LocalPatricFam, etc.)


=item Definition

=begin html

<pre>
a reference to a list containing 4 items:
0: (db) a string
1: (id) a string
2: (function) a string
3: (db_version) a string

</pre>

=end html

=begin text

a reference to a list containing 4 items:
0: (db) a string
1: (id) a string
2: (function) a string
3: (db_version) a string


=end text

=back



=head2 similarity_association

=over 4



=item Description

* A similarity association notes the BLAST-computed association
* between this feature and a given protein database.


=item Definition

=begin html

<pre>
a reference to a list containing 6 items:
0: (source) a string
1: (source_id) a string
2: (query_coverage) a float
3: (subject_coverage) a float
4: (identity) a float
5: (e_value) a float

</pre>

=end html

=begin text

a reference to a list containing 6 items:
0: (source) a string
1: (source_id) a string
2: (query_coverage) a float
3: (subject_coverage) a float
4: (identity) a float
5: (e_value) a float


=end text

=back



=head2 proposed_function

=over 4



=item Description

A proposed function records an assertion of the function of a feature.
* A feature may have multiple proposed functions. A tool downstream of the
* tools that propose functions may determine based on the asserted proposals
* which function should be the assigned function for the feature.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is a string
function has a value which is a string
user has a value which is a string
score has a value which is a float
event_id has a value which is an analysis_event_id
timestamp has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is a string
function has a value which is a string
user has a value which is a string
score has a value which is a float
event_id has a value which is an analysis_event_id
timestamp has a value which is an int


=end text

=back



=head2 genbank_feature

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
genbank_type has a value which is a string
genbank_location has a value which is a string
values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
genbank_type has a value which is a string
genbank_location has a value which is a string
values has a value which is a reference to a hash where the key is a string and the value is a reference to a list where each element is a string


=end text

=back



=head2 genbank_locus

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
accession has a value which is a reference to a list where each element is a string
comment has a value which is a reference to a list where each element is a string
date has a value which is a string
dblink has a value which is a reference to a list where each element is a string
dbsource has a value which is a reference to a list where each element is a string
definition has a value which is a string
division has a value which is a string
geometry has a value which is a string
gi has a value which is an int
keywords has a value which is a reference to a list where each element is a string
locus has a value which is a string
organism has a value which is a string
origin has a value which is a string
references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
source has a value which is a string
taxonomy has a value which is a reference to a list where each element is a string
version has a value which is a reference to a list where each element is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
accession has a value which is a reference to a list where each element is a string
comment has a value which is a reference to a list where each element is a string
date has a value which is a string
dblink has a value which is a reference to a list where each element is a string
dbsource has a value which is a reference to a list where each element is a string
definition has a value which is a string
division has a value which is a string
geometry has a value which is a string
gi has a value which is an int
keywords has a value which is a reference to a list where each element is a string
locus has a value which is a string
organism has a value which is a string
origin has a value which is a string
references has a value which is a reference to a list where each element is a reference to a hash where the key is a string and the value is a string
source has a value which is a string
taxonomy has a value which is a reference to a list where each element is a string
version has a value which is a reference to a list where each element is a string


=end text

=back



=head2 feature

=over 4



=item Description

* The function_id refers to the particular proposed function that was chosen
* for this feature.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is a feature_id
location has a value which is a location
type has a value which is a feature_type
function has a value which is a string
function_id has a value which is a string
protein_translation has a value which is a string
aliases has a value which is a reference to a list where each element is a string
alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (source) a string
1: (alias) a string

annotations has a value which is a reference to a list where each element is an annotation
quality has a value which is a feature_quality_measure
feature_creation_event has a value which is an analysis_event_id
family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
similarity_associations has a value which is a reference to a list where each element is a similarity_association
proposed_functions has a value which is a reference to a list where each element is a proposed_function
genbank_type has a value which is a string
genbank_feature has a value which is a genbank_feature

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is a feature_id
location has a value which is a location
type has a value which is a feature_type
function has a value which is a string
function_id has a value which is a string
protein_translation has a value which is a string
aliases has a value which is a reference to a list where each element is a string
alias_pairs has a value which is a reference to a list where each element is a reference to a list containing 2 items:
0: (source) a string
1: (alias) a string

annotations has a value which is a reference to a list where each element is an annotation
quality has a value which is a feature_quality_measure
feature_creation_event has a value which is an analysis_event_id
family_assignments has a value which is a reference to a list where each element is a protein_family_assignment
similarity_associations has a value which is a reference to a list where each element is a similarity_association
proposed_functions has a value which is a reference to a list where each element is a proposed_function
genbank_type has a value which is a string
genbank_feature has a value which is a genbank_feature


=end text

=back



=head2 contig

=over 4



=item Description

circular / linear


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is a contig_id
dna has a value which is a string
genetic_code has a value which is an int
cell_compartment has a value which is a string
replicon_type has a value which is a string
replicon_geometry has a value which is a string
complete has a value which is a bool
genbank_locus has a value which is a genbank_locus

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is a contig_id
dna has a value which is a string
genetic_code has a value which is an int
cell_compartment has a value which is a string
replicon_type has a value which is a string
replicon_geometry has a value which is a string
complete has a value which is a bool
genbank_locus has a value which is a genbank_locus


=end text

=back



=head2 close_genome

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
genome_name has a value which is a string
closeness_measure has a value which is a float
analysis_method has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
genome has a value which is a genome_id
genome_name has a value which is a string
closeness_measure has a value which is a float
analysis_method has a value which is a string


=end text

=back



=head2 genome_quality_measure

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
frameshift_error_rate has a value which is a float
sequence_error_rate has a value which is a float

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
frameshift_error_rate has a value which is a float
sequence_error_rate has a value which is a float


=end text

=back



=head2 strain_type

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
typing_method has a value which is a string
database has a value which is a string
tag has a value which is a string
event_id has a value which is an analysis_event_id

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
typing_method has a value which is a string
database has a value which is a string
tag has a value which is a string
event_id has a value which is an analysis_event_id


=end text

=back



=head2 classifier

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
version has a value which is a string
description has a value which is a string
comment has a value which is a string
antibiotics has a value which is a reference to a list where each element is a string
accuracy has a value which is a float
area_under_roc_curve has a value which is a float
f1_score has a value which is a float
sources has a value which is a string
cumulative_adaboost_value has a value which is a float
sensitivity has a value which is a string
event_id has a value which is an analysis_event_id
features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
0: (id) a feature_id
1: (alpha) a float
2: (round) an int
3: (function) a string


</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
name has a value which is a string
version has a value which is a string
description has a value which is a string
comment has a value which is a string
antibiotics has a value which is a reference to a list where each element is a string
accuracy has a value which is a float
area_under_roc_curve has a value which is a float
f1_score has a value which is a float
sources has a value which is a string
cumulative_adaboost_value has a value which is a float
sensitivity has a value which is a string
event_id has a value which is an analysis_event_id
features has a value which is a reference to a list where each element is a reference to a list containing 4 items:
0: (id) a feature_id
1: (alpha) a float
2: (round) an int
3: (function) a string



=end text

=back



=head2 genomeTO

=over 4



=item Description

All of the information about particular genome


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is a genome_id
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
source has a value which is a string
source_id has a value which is a string
taxonomy has a value which is a string
ncbi_taxonomy_id has a value which is an int
owner has a value which is a string
quality has a value which is a genome_quality_measure
contigs has a value which is a reference to a list where each element is a contig
contigs_handle has a value which is a Handle
features has a value which is a reference to a list where each element is a feature
close_genomes has a value which is a reference to a list where each element is a close_genome
analysis_events has a value which is a reference to a list where each element is an analysis_event
typing has a value which is a reference to a list where each element is a strain_type
classifications has a value which is a reference to a list where each element is a classifier

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is a genome_id
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
source has a value which is a string
source_id has a value which is a string
taxonomy has a value which is a string
ncbi_taxonomy_id has a value which is an int
owner has a value which is a string
quality has a value which is a genome_quality_measure
contigs has a value which is a reference to a list where each element is a contig
contigs_handle has a value which is a Handle
features has a value which is a reference to a list where each element is a feature
close_genomes has a value which is a reference to a list where each element is a close_genome
analysis_events has a value which is a reference to a list where each element is an analysis_event
typing has a value which is a reference to a list where each element is a strain_type
classifications has a value which is a reference to a list where each element is a classifier


=end text

=back



=head2 genome_metadata

=over 4



=item Description

* Genome metadata. We use this structure to define common metadata
* settings used in the API calls below. It is possible this data should
* have been separated in this way in the genome object itself, but there
* is an extant body of code that assumes the current structure of the genome
* object.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
id has a value which is a genome_id
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
source has a value which is a string
source_id has a value which is a string
ncbi_taxonomy_id has a value which is an int
taxonomy has a value which is a string
owner has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
id has a value which is a genome_id
scientific_name has a value which is a string
domain has a value which is a string
genetic_code has a value which is an int
source has a value which is a string
source_id has a value which is a string
ncbi_taxonomy_id has a value which is an int
taxonomy has a value which is a string
owner has a value which is a string


=end text

=back



=head2 subsystem

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 variant

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 variant_of_subsystem

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 2 items:
0: a subsystem
1: a variant

</pre>

=end html

=begin text

a reference to a list containing 2 items:
0: a subsystem
1: a variant


=end text

=back



=head2 variant_subsystem_pairs

=over 4



=item Definition

=begin html

<pre>
a reference to a list where each element is a variant_of_subsystem
</pre>

=end html

=begin text

a reference to a list where each element is a variant_of_subsystem

=end text

=back



=head2 fid

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 role

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 function

=over 4



=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 fid_role_pair

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 2 items:
0: a fid
1: a role

</pre>

=end html

=begin text

a reference to a list containing 2 items:
0: a fid
1: a role


=end text

=back



=head2 fid_role_pairs

=over 4



=item Definition

=begin html

<pre>
a reference to a list where each element is a fid_role_pair
</pre>

=end html

=begin text

a reference to a list where each element is a fid_role_pair

=end text

=back



=head2 fid_function_pair

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 2 items:
0: a fid
1: a function

</pre>

=end html

=begin text

a reference to a list containing 2 items:
0: a fid
1: a function


=end text

=back



=head2 fid_function_pairs

=over 4



=item Definition

=begin html

<pre>
a reference to a list where each element is a fid_function_pair
</pre>

=end html

=begin text

a reference to a list where each element is a fid_function_pair

=end text

=back



=head2 reconstructionTO

=over 4



=item Description

Metabolic reconstruction
represents the set of subsystems that we infer are present in this genome


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
subsystems has a value which is a variant_subsystem_pairs
bindings has a value which is a fid_role_pairs
assignments has a value which is a fid_function_pairs

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
subsystems has a value which is a variant_subsystem_pairs
bindings has a value which is a fid_role_pairs
assignments has a value which is a fid_function_pairs


=end text

=back



=head2 fid_data_tuple

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 4 items:
0: a fid
1: a md5
2: a location
3: a function

</pre>

=end html

=begin text

a reference to a list containing 4 items:
0: a fid
1: a md5
2: a location
3: a function


=end text

=back



=head2 fid_data_tuples

=over 4



=item Definition

=begin html

<pre>
a reference to a list where each element is a fid_data_tuple
</pre>

=end html

=begin text

a reference to a list where each element is a fid_data_tuple

=end text

=back



=head2 compact_feature

=over 4



=item Description

* This tuple defines a compact form for defining features to be batch-loaded
* into a genome object.


=item Definition

=begin html

<pre>
a reference to a list containing 5 items:
0: (id) a string
1: (location) a string
2: (feature_type) a string
3: (function) a string
4: (aliases) a string

</pre>

=end html

=begin text

a reference to a list containing 5 items:
0: (id) a string
1: (location) a string
2: (feature_type) a string
3: (function) a string
4: (aliases) a string


=end text

=back



=head2 rna_type

=over 4



=item Description

[ validate.enum("5S", "SSU", "LSU", "ALL") ]


=item Definition

=begin html

<pre>
a string
</pre>

=end html

=begin text

a string

=end text

=back



=head2 glimmer3_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
min_training_len has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
min_training_len has a value which is an int


=end text

=back



=head2 SEED_projection_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
reference_database has a value which is a string
reference_id has a value which is a string
kmer_size has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
reference_database has a value which is a string
reference_id has a value which is a string
kmer_size has a value which is an int


=end text

=back



=head2 repeat_region_SEED_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
min_identity has a value which is a float
min_length has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
min_identity has a value which is a float
min_length has a value which is an int


=end text

=back



=head2 assembly_gap_parameters

=over 4



=item Description

* Given a genome typed object, call gap features.
* Gaps are known regions in the contig where the nucleotide sequence is not known
* but where there is evidence that a run of DNA does exist joining the sequenced
* data on either side of the gap.
*
* Gaps are currently called using one of two methods. Genomes that originated as
* genbank files may have a CONTIGS entry that defines the contig and gap regions.
* Genomes that do not have a CONTIGS entry are scanned for runs of "n" characters.


=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
min_gap_length has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
min_gap_length has a value which is an int


=end text

=back



=head2 similarity_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
annotate_hypothetical_only has a value which is an int


=end text

=back



=head2 phage_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
annotate_hypothetical_only has a value which is an int


=end text

=back



=head2 kmer_v1_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
kmer_size has a value which is an int
dataset_name has a value which is a string
return_scores_for_all_proteins has a value which is an int
score_threshold has a value which is an int
hit_threshold has a value which is an int
sequential_hit_threshold has a value which is an int
detailed has a value which is an int
min_hits has a value which is an int
min_size has a value which is an int
max_gap has a value which is an int
annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
kmer_size has a value which is an int
dataset_name has a value which is a string
return_scores_for_all_proteins has a value which is an int
score_threshold has a value which is an int
hit_threshold has a value which is an int
sequential_hit_threshold has a value which is an int
detailed has a value which is an int
min_hits has a value which is an int
min_size has a value which is an int
max_gap has a value which is an int
annotate_hypothetical_only has a value which is an int


=end text

=back



=head2 kmer_v2_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
min_hits has a value which is an int
max_gap has a value which is an int
annotate_hypothetical_only has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
min_hits has a value which is an int
max_gap has a value which is an int
annotate_hypothetical_only has a value which is an int


=end text

=back



=head2 resolve_overlapping_features_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
placeholder has a value which is an int

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
placeholder has a value which is an int


=end text

=back



=head2 propagate_genbank_feature_metadata_parameters

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
min_rna_pct_coverage has a value which is a float

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
min_rna_pct_coverage has a value which is a float


=end text

=back



=head2 special_protein_hit

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 7 items:
0: (protein_id) a string
1: (database_name) a string
2: (database_id) a string
3: (protein_coverage) a string
4: (database_coverage) a string
5: (identity) a float
6: (p_value) a float

</pre>

=end html

=begin text

a reference to a list containing 7 items:
0: (protein_id) a string
1: (database_name) a string
2: (database_id) a string
3: (protein_coverage) a string
4: (database_coverage) a string
5: (identity) a float
6: (p_value) a float


=end text

=back



=head2 cdd_hit

=over 4



=item Definition

=begin html

<pre>
a reference to a list containing 16 items:
0: (protein_id) a string
1: (domain_id) a string
2: (identity) a float
3: (alignment_len) an int
4: (mismatches) an int
5: (gap_openings) an int
6: (protein_start) an int
7: (protein_end) an int
8: (domain_start) an int
9: (domain_end) an int
10: (e_value) a float
11: (bit_score) a float
12: (accession) a string
13: (short_name) a string
14: (description) a string
15: (pssm_length) an int

</pre>

=end html

=begin text

a reference to a list containing 16 items:
0: (protein_id) a string
1: (domain_id) a string
2: (identity) a float
3: (alignment_len) an int
4: (mismatches) an int
5: (gap_openings) an int
6: (protein_start) an int
7: (protein_end) an int
8: (domain_start) an int
9: (domain_end) an int
10: (e_value) a float
11: (bit_score) a float
12: (accession) a string
13: (short_name) a string
14: (description) a string
15: (pssm_length) an int


=end text

=back



=head2 pipeline_stage

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
name has a value which is a string
condition has a value which is a string
failure_is_not_fatal has a value which is an int
repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
glimmer3_parameters has a value which is a glimmer3_parameters
kmer_v1_parameters has a value which is a kmer_v1_parameters
kmer_v2_parameters has a value which is a kmer_v2_parameters
similarity_parameters has a value which is a similarity_parameters

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
name has a value which is a string
condition has a value which is a string
failure_is_not_fatal has a value which is an int
repeat_region_SEED_parameters has a value which is a repeat_region_SEED_parameters
glimmer3_parameters has a value which is a glimmer3_parameters
kmer_v1_parameters has a value which is a kmer_v1_parameters
kmer_v2_parameters has a value which is a kmer_v2_parameters
similarity_parameters has a value which is a similarity_parameters


=end text

=back



=head2 workflow

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
stages has a value which is a reference to a list where each element is a pipeline_stage

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
stages has a value which is a reference to a list where each element is a pipeline_stage


=end text

=back



=head2 pipeline_batch_input

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
genome_id has a value which is a string
data has a value which is a Handle
filename has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
genome_id has a value which is a string
data has a value which is a Handle
filename has a value which is a string


=end text

=back



=head2 pipeline_batch_status_entry

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
genome_id has a value which is a string
status has a value which is a string
creation_date has a value which is a string
start_date has a value which is a string
completion_date has a value which is a string
stdout has a value which is a Handle
stderr has a value which is a Handle
output has a value which is a Handle
filename has a value which is a string

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
genome_id has a value which is a string
status has a value which is a string
creation_date has a value which is a string
start_date has a value which is a string
completion_date has a value which is a string
stdout has a value which is a Handle
stderr has a value which is a Handle
output has a value which is a Handle
filename has a value which is a string


=end text

=back



=head2 pipeline_batch_status

=over 4



=item Definition

=begin html

<pre>
a reference to a hash where the following keys are defined:
status has a value which is a string
submit_date has a value which is a string
start_date has a value which is a string
completion_date has a value which is a string
details has a value which is a reference to a list where each element is a pipeline_batch_status_entry

</pre>

=end html

=begin text

a reference to a hash where the following keys are defined:
status has a value which is a string
submit_date has a value which is a string
start_date has a value which is a string
completion_date has a value which is a string
details has a value which is a reference to a list where each element is a pipeline_batch_status_entry


=end text

=back



=cut

1;
